

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>scspecies.preprocessing &mdash; scspecies Documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../_static/jupyter-sphinx.css?v=572af1d6" />
      <link rel="stylesheet" type="text/css" href="../../_static/thebelab.css" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=01f34227"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../../_static/thebelab-helper.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            scspecies
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Navigation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/api.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tips.html">Tips to fit scSpecies</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">scspecies</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">scspecies.preprocessing</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for scspecies.preprocessing</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">List</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">collections</span><span class="w"> </span><span class="kn">import</span> <span class="n">Counter</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm.auto</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">muon</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mu</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">h5py</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">io</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">tempfile</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">requests</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.sparse</span><span class="w"> </span><span class="kn">import</span> <span class="n">csr_matrix</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">mygene</span><span class="w"> </span><span class="kn">import</span> <span class="n">MyGeneInfo</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">contextlib</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">OneHotEncoder</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.neighbors</span><span class="w"> </span><span class="kn">import</span> <span class="n">NearestNeighbors</span>


<div class="viewcode-block" id="map_homologs">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.map_homologs">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">map_homologs</span><span class="p">(</span>
        <span class="n">gene_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> 
        <span class="n">target_NCBI_Taxon_ID</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
        <span class="n">context_NCBI_Taxon_ID</span><span class="p">:</span> <span class="nb">int</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Maps a list of gene symbols from the target species to their homologous symbols</span>
<span class="sd">    of the context species using MyGeneInfo.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene_list : list[str]</span>
<span class="sd">        Gene symbols in the target species to be translated.</span>
<span class="sd">    target_NCBI_Taxon_ID : int</span>
<span class="sd">        NCBI Taxonomy ID of the target species.</span>
<span class="sd">    context_NCBI_Taxon_ID : int</span>
<span class="sd">        NCBI Taxonomy ID of the source (context) species.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list[str]</span>
<span class="sd">        Homologous gene symbols in the target species, with &#39;non_hom_&lt;i&gt;&#39; for non homologous genes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">mg</span> <span class="o">=</span> <span class="n">MyGeneInfo</span><span class="p">()</span>
    
    <span class="n">results</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span>
        <span class="n">gene_list</span><span class="p">,</span>
        <span class="n">scopes</span><span class="o">=</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span>
        <span class="n">species</span><span class="o">=</span><span class="n">target_NCBI_Taxon_ID</span><span class="p">,</span>
        <span class="n">fields</span><span class="o">=</span><span class="s1">&#39;homologene&#39;</span><span class="p">,</span>
        <span class="n">as_dataframe</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>

    <span class="n">homolog_ids</span> <span class="o">=</span> <span class="p">{}</span>       
    <span class="n">ids_to_lookup</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>  
    <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="n">src</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;notfound&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="s1">&#39;homologene&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
            <span class="n">homolog_ids</span><span class="p">[</span><span class="n">src</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">hits</span> <span class="o">=</span> <span class="p">[</span><span class="n">g</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">res</span><span class="p">[</span><span class="s1">&#39;homologene&#39;</span><span class="p">][</span><span class="s1">&#39;genes&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">g</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">context_NCBI_Taxon_ID</span><span class="p">]</span>
            <span class="n">homolog_ids</span><span class="p">[</span><span class="n">src</span><span class="p">]</span> <span class="o">=</span> <span class="n">hits</span>
            <span class="n">ids_to_lookup</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">hits</span><span class="p">)</span>

    <span class="n">id_to_symbol</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">ids_to_lookup</span><span class="p">:</span>
        <span class="n">lookup</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span><span class="n">ids_to_lookup</span><span class="p">),</span>
            <span class="n">scopes</span><span class="o">=</span><span class="s1">&#39;entrezgene&#39;</span><span class="p">,</span>
            <span class="n">fields</span><span class="o">=</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span>
            <span class="n">species</span><span class="o">=</span><span class="n">context_NCBI_Taxon_ID</span><span class="p">,</span>
            <span class="n">as_dataframe</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
        
        <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">lookup</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">eid</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">hit</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">])</span>
                <span class="k">if</span> <span class="s1">&#39;symbol&#39;</span> <span class="ow">in</span> <span class="n">hit</span><span class="p">:</span>
                    <span class="n">id_to_symbol</span><span class="p">[</span><span class="n">eid</span><span class="p">]</span> <span class="o">=</span> <span class="n">hit</span><span class="p">[</span><span class="s1">&#39;symbol&#39;</span><span class="p">]</span>
            <span class="k">except</span> <span class="p">(</span><span class="ne">KeyError</span><span class="p">,</span> <span class="ne">ValueError</span><span class="p">):</span>
                <span class="k">continue</span>

    <span class="n">mapped</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">g</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_list</span><span class="p">):</span>
        <span class="n">syms</span> <span class="o">=</span> <span class="p">[</span><span class="n">id_to_symbol</span><span class="p">[</span><span class="n">eid</span><span class="p">]</span> <span class="k">for</span> <span class="n">eid</span> <span class="ow">in</span> <span class="n">homolog_ids</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="p">[])</span> <span class="k">if</span> <span class="n">eid</span> <span class="ow">in</span> <span class="n">id_to_symbol</span><span class="p">]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">syms</span><span class="p">:</span>
            <span class="n">mapped</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;non_hom_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">mapped</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">syms</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">mapped</span></div>


<div class="viewcode-block" id="map_homologs_silent">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.map_homologs_silent">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">map_homologs_silent</span><span class="p">(</span>
        <span class="n">gene_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> 
        <span class="n">target_NCBI_Taxon_ID</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
        <span class="n">context_NCBI_Taxon_ID</span><span class="p">:</span> <span class="nb">int</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Same as `map_homologs` but suppresses all console output </span>
<span class="sd">    as map_homologs outputs a print statement for each gene.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene_list : list[str]</span>
<span class="sd">        Gene symbols in the target species to be translated.</span>
<span class="sd">    target_NCBI_Taxon_ID : int</span>
<span class="sd">        NCBI Taxonomy ID of the target species.</span>
<span class="sd">    context_NCBI_Taxon_ID : int</span>
<span class="sd">        NCBI Taxonomy ID of the source (context) species.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list[str]</span>
<span class="sd">        Homologous gene symbols in the target species, with &#39;non_hom_&lt;i&gt;&#39; for non homologous genes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">buf</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
    <span class="k">with</span> <span class="n">contextlib</span><span class="o">.</span><span class="n">redirect_stdout</span><span class="p">(</span><span class="n">buf</span><span class="p">),</span> <span class="n">contextlib</span><span class="o">.</span><span class="n">redirect_stderr</span><span class="p">(</span><span class="n">buf</span><span class="p">):</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">map_homologs</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">target_NCBI_Taxon_ID</span><span class="p">,</span> <span class="n">context_NCBI_Taxon_ID</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="get_key">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.get_key">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_key</span><span class="p">(</span>
        <span class="n">gene</span><span class="p">,</span> 
        <span class="n">homology_targsp_df</span><span class="p">,</span> 
        <span class="n">homology_context_df</span><span class="p">,</span> 
        <span class="n">i</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span> 
<span class="w">    </span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve the homologous context gene symbol for a given target gene using homology tables from</span>
<span class="sd">    informatics.jax.org/downloads/reports/HOM_AllOrganism.rpt</span>
<span class="sd">    Can only be used for mouse, rat, human, zebrafish context-target dataset pairs,</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene : str</span>
<span class="sd">        Gene symbol in the ‘from’ DataFrame.</span>
<span class="sd">    homology_targsp_df : pandas.DataFrame</span>
<span class="sd">        Homology table for the target species (columns include &#39;Symbol&#39; and &#39;DB Class Key&#39;).</span>
<span class="sd">    homology_context_df : pandas.DataFrame</span>
<span class="sd">        Homology table for the context species (same key column).</span>
<span class="sd">    i : int</span>
<span class="sd">        Index of the gene in the original list, used to name unmapped genes.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Context‐species gene symbol if found, otherwise &#39;non_hom_&lt;i&gt;&#39;.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">targ_gene_names</span> <span class="o">=</span> <span class="s1">&#39;non_hom_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">homology_targsp_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">key</span> <span class="o">=</span> <span class="n">homology_targsp_df</span><span class="p">[</span><span class="n">homology_targsp_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">][</span><span class="s1">&#39;DB Class Key&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> 

        <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">homology_context_df</span><span class="p">[</span><span class="s1">&#39;DB Class Key&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="n">targ_gene_names</span> <span class="o">=</span> <span class="n">homology_context_df</span><span class="p">[</span><span class="n">homology_context_df</span><span class="p">[</span><span class="s1">&#39;DB Class Key&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">key</span><span class="p">][</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">targ_gene_names</span>      </div>



<div class="viewcode-block" id="download_datasets">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.download_datasets">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">download_datasets</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download liver cell .h5ad datasets into ./data directory.</span>
<span class="sd">    Downloads each file and skips files already present.</span>
<span class="sd">    </span>
<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    requests.HTTPError</span>
<span class="sd">        If any of the dataset URLs returns a bad status.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_urls</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;human_liver.h5ad&quot;</span><span class="p">:</span>   <span class="s2">&quot;https://zenodo.org/records/15522251/files/human_liver.h5ad?download=1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;mouse_liver.h5ad&quot;</span><span class="p">:</span>   <span class="s2">&quot;https://zenodo.org/records/15522251/files/mouse_liver.h5ad?download=1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;hamster_liver.h5ad&quot;</span><span class="p">:</span> <span class="s2">&quot;https://zenodo.org/records/15522251/files/hamster_liver.h5ad?download=1&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">data_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;data&quot;</span><span class="p">)</span>
    <span class="n">data_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fname</span><span class="p">,</span> <span class="n">url</span> <span class="ow">in</span> <span class="n">data_urls</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">out_file</span> <span class="o">=</span> <span class="n">data_path</span> <span class="o">/</span> <span class="n">fname</span>
        <span class="k">if</span> <span class="n">out_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2"> already exists. Skipping download.&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Downloading </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2"> …&quot;</span><span class="p">)</span>
        <span class="n">resp</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>  <span class="c1"># no stream=True</span>
        <span class="n">resp</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">resp</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>

        <span class="n">size_mb</span> <span class="o">=</span> <span class="n">out_file</span><span class="o">.</span><span class="n">stat</span><span class="p">()</span><span class="o">.</span><span class="n">st_size</span> <span class="o">/</span> <span class="mi">1024</span> <span class="o">/</span> <span class="mi">1024</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2"> downloaded. Size: </span><span class="si">{</span><span class="n">size_mb</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> MB&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;All datasets have been downloaded to the ./data directory.&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="set_random_seed">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.set_random_seed">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">set_random_seed</span><span class="p">(</span>
        <span class="n">seed</span><span class="p">:</span> <span class="nb">int</span>
        <span class="p">):</span>
<span class="w">    </span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fix all relevant RNG seeds for reproducibility.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seed : int</span>
<span class="sd">        The seed value to use for Python, NumPy, random, and PyTorch.</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;PYTHONHASHSEED&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span> 
    <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span> 
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>  
    <span class="n">torch</span><span class="o">.</span><span class="n">manual_seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>  
    <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">():</span>
        <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">manual_seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>  
        <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">manual_seed_all</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>  
        <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">deterministic</span> <span class="o">=</span> <span class="kc">True</span> 
        <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">benchmark</span> <span class="o">=</span> <span class="kc">False</span></div>


<div class="viewcode-block" id="create_mdata">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">create_mdata</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Builder for MuData container that is used by scSpecies to align context &amp; target AnnData datasets.</span>

<span class="sd">    Handles downloading a gene-translation table from the mouse to human genome, </span>
<span class="sd">    preprocessing a “context” AnnData, and “target” AnnData from potentially multiple species, and saving the</span>
<span class="sd">    final MuData object.</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="create_mdata.__init__">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.__init__">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>        
            <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
            <span class="n">batch_key</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">cell_key</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>      
            <span class="n">dataset_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;mouse&#39;</span><span class="p">,</span>
            <span class="n">NCBI_Taxon_ID</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10090</span><span class="p">,</span>
            <span class="n">n_top_genes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>                   
            <span class="n">min_non_zero_genes</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.025</span><span class="p">,</span> 
            <span class="n">min_cell_type_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>     
            <span class="n">min_batch_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>                
            <span class="p">):</span> 

<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize and preprocess the context dataset.</span>

<span class="sd">        Steps:</span>

<span class="sd">        1. Onehot-encode experimental batchs.</span>

<span class="sd">        2. Calculate library size encoder prior parameters for scVI</span>

<span class="sd">        3. Subset to top HVGs and filter out cells with low expression patterns as well as rare cell types and batches (optionally).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : ad.AnnData</span>
<span class="sd">            AnnData used as a context in scSpecies.</span>
<span class="sd">        batch_key : str</span>
<span class="sd">            Observation key for experimental batch labels.</span>
<span class="sd">        cell_key : str</span>
<span class="sd">            Observation key for cell-type annotation.</span>
<span class="sd">        dataset_name : str, optional</span>
<span class="sd">            Tag for the context dataset (default &#39;mouse&#39;).</span>
<span class="sd">        NCBI_Taxon_ID : int, optional</span>
<span class="sd">            Taxonomy ID of the context species (default mouse - 10090).</span>
<span class="sd">        n_top_genes : int or None, optional</span>
<span class="sd">            Number of HVGs to retain (None to skip) (default None).</span>
<span class="sd">        min_non_zero_genes : float, optional</span>
<span class="sd">            Min fraction of nonzero genes per cell (default 0.025).</span>
<span class="sd">        min_cell_type_size : int, optional</span>
<span class="sd">            Min cells per cell-type, cell types with fewer samples are removed (default 20).</span>
<span class="sd">        min_batch_size : int, optional</span>
<span class="sd">            Min cells per batch for encoding, batch with fewer samples are removed (default 20).</span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - Ensures a `data/` directory exists.</span>
<span class="sd">        - Annotates `adata.uns[&#39;metadata&#39;]` with context dataset info.</span>
<span class="sd">        - One-hot encodes batch labels, dropping any batches smaller than `min_batch_size`.</span>
<span class="sd">        - Computes per-batch library size prior parameters.</span>
<span class="sd">        - Subsets to top highly variable genes if `n_top_genes` is not None.</span>
<span class="sd">        - Filters out cells with low gene detection and rare cell-types.</span>
<span class="sd">        - Stores the processed AnnData in `self.dataset_collection`.</span>
<span class="sd">        &quot;&quot;&quot;</span>        

        <span class="n">mu</span><span class="o">.</span><span class="n">set_options</span><span class="p">(</span><span class="n">pull_on_update</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">min_non_zero_genes</span> <span class="o">=</span> <span class="n">min_non_zero_genes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">min_cell_type_size</span> <span class="o">=</span> <span class="n">min_cell_type_size</span>  
        <span class="bp">self</span><span class="o">.</span><span class="n">min_batch_size</span> <span class="o">=</span> <span class="n">min_batch_size</span>   
        
        <span class="n">out_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;data&quot;</span><span class="p">)</span>
        <span class="n">out_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">context_dataset_name</span> <span class="o">=</span> <span class="n">dataset_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">context_NCBI_Taxon_ID</span> <span class="o">=</span> <span class="n">NCBI_Taxon_ID</span>
        
        <span class="k">if</span> <span class="n">adata</span><span class="o">.</span><span class="n">isbacked</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="n">dataset_name</span><span class="p">,</span>            
            <span class="s1">&#39;batch_key&#39;</span><span class="p">:</span> <span class="n">batch_key</span><span class="p">,</span> 
            <span class="s1">&#39;cell_key&#39;</span><span class="p">:</span> <span class="n">cell_key</span><span class="p">,</span>
            <span class="s1">&#39;NCBI_Taxon_ID&#39;</span><span class="p">:</span> <span class="n">NCBI_Taxon_ID</span><span class="p">,</span>
            <span class="s1">&#39;function&#39;</span><span class="p">:</span> <span class="s1">&#39;context&#39;</span><span class="p">,</span>            
        <span class="p">}</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;dataset&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataset_name</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">&quot;</span>        

        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">encode_batch_labels</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_batch_size</span><span class="p">)</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_library_prior_params</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>     
        
        <span class="k">if</span> <span class="n">n_top_genes</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">subset_to_highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="p">)</span>

        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_non_zero_genes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_cell_type_size</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">dataset_collection</span> <span class="o">=</span> <span class="p">{</span><span class="n">dataset_name</span><span class="p">:</span> <span class="n">adata</span><span class="p">}</span>
        
        <span class="n">adata</span><span class="o">.</span><span class="n">obs_names_make_unique</span><span class="p">()</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done!</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">+</span><span class="s1">&#39;-&#39;</span><span class="o">*</span><span class="mi">90</span><span class="p">)</span></div>


<div class="viewcode-block" id="create_mdata.setup_target_adata">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.setup_target_adata">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">setup_target_adata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
            <span class="n">batch_key</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">cell_key</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>  
            <span class="n">eval_nns_keys</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">dataset_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">,</span>
            <span class="n">NCBI_Taxon_ID</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">n_top_genes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>               
            <span class="n">compute_log1p</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>                   
            <span class="n">nn_kwargs</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>     
            <span class="p">):</span>   

<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Preprocess and align a target AnnData against the context.</span>

<span class="sd">        Steps:</span>

<span class="sd">        1. Onehot-encode experimental batchs.</span>

<span class="sd">        2. Calculate library size encoder prior parameters for scVI</span>

<span class="sd">        3. Subset to top HVGs and filter out cells with low expression patterns as well as rare cell types and batches (optionally). </span>

<span class="sd">        4. Translate target gene symbols to context homologs.</span>

<span class="sd">        5. Compute and evaluate data-level nearest neighbors on the shared homologous gene set.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : ad.AnnData</span>
<span class="sd">            Target dataset.</span>
<span class="sd">        batch_key : str</span>
<span class="sd">            Observation key for experimental batch labels.</span>
<span class="sd">        cell_key : str or None</span>
<span class="sd">            Observation key for cell types (None if unkown).</span>
<span class="sd">        eval_nns_keys : List of str or None</span>
<span class="sd">            List of context dataset `obs` keys that should be transferred by scSpecies. Defaults to [cell_key]. </span>
<span class="sd">        dataset_name : str, optional</span>
<span class="sd">            Defaults to &#39;human&#39;.</span>
<span class="sd">        NCBI_Taxon_ID : int, optional</span>
<span class="sd">            Taxonomy ID for the target species (default human - 9606).</span>
<span class="sd">        n_top_genes : int or None, optional</span>
<span class="sd">            Number of HVGs to keep (None to skip) (default None).</span>
<span class="sd">        compute_log1p : bool, optional</span>
<span class="sd">            Use log1p counts for neighbor search if True (default True).</span>
<span class="sd">        nn_kwargs : dict, optional</span>
<span class="sd">            Args for sklearn.neighbors.NearestNeighbors.</span>
<span class="sd">            Defaults to `{&#39;n_neighbors&#39;: 250, &#39;metric&#39;: &#39;cosine&#39;}`.       </span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - Updates `adata.uns[&#39;metadata&#39;]` with target dataset info.</span>
<span class="sd">        - Filters and one-hot encodes batch (and cell-type, if provided).</span>
<span class="sd">        - Computes library size prior parameters.</span>
<span class="sd">        - Calls `translate_gene_list` to add translated gene symbols in the context genome to `var_names_transl`.</span>
<span class="sd">        - Subsets to HVGs if `n_top_genes` is not None.</span>
<span class="sd">        - Filters out low-coverage cells and rare cell-types.</span>
<span class="sd">        - Identifies intersecting homologous genes with the context and performs a nearest-neighbor search on log1p (or raw) counts.</span>
<span class="sd">        - Stores neighbor indices in `adata.obsm[&#39;ind_neigh_nns&#39;]`.</span>
<span class="sd">        - Calculates the percentage of neighbor label agreement and transfers labels based on the data-level nearest neighbor search.</span>
<span class="sd">        - Inserts the processed AnnData into `self.dataset_collection`.        </span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="n">dataset_name</span><span class="p">,</span>            
            <span class="s1">&#39;batch_key&#39;</span><span class="p">:</span> <span class="n">batch_key</span><span class="p">,</span>
            <span class="s1">&#39;cell_key&#39;</span><span class="p">:</span> <span class="n">cell_key</span><span class="p">,</span> 
            <span class="s1">&#39;NCBI_Taxon_ID&#39;</span><span class="p">:</span> <span class="n">NCBI_Taxon_ID</span><span class="p">,</span>
            <span class="s1">&#39;function&#39;</span><span class="p">:</span> <span class="s1">&#39;target&#39;</span><span class="p">,</span>                  
        <span class="p">}</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;dataset&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataset_name</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">&quot;</span>        

        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">encode_batch_labels</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_batch_size</span><span class="p">)</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_library_prior_params</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>           

        <span class="k">if</span> <span class="n">n_top_genes</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">subset_to_highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="p">)</span>

        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_non_zero_genes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_cell_type_size</span><span class="p">)</span>

        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_gene_list</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">context_ind</span><span class="p">,</span> <span class="n">target_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataset_collection</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">context_dataset_name</span><span class="p">]</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;var_names_transl&#39;</span><span class="p">],</span> <span class="n">return_indices</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">nn_kwargs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">nn_kwargs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="s2">&quot;n_neighbors&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">nn_kwargs</span><span class="p">:</span>
            <span class="n">nn_kwargs</span><span class="p">[</span><span class="s2">&quot;n_neighbors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">250</span>
        <span class="k">if</span> <span class="s2">&quot;metric&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">nn_kwargs</span><span class="p">:</span>
            <span class="n">nn_kwargs</span><span class="p">[</span><span class="s2">&quot;metric&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;cosine&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">context_ind</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No homologous genes found. scSpecies cannot be used.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">context_ind</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">250</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Warning</span><span class="p">(</span><span class="s2">&quot;Only </span><span class="se">\033</span><span class="s2">[35m</span><span class="si">{}</span><span class="se">\033</span><span class="s2">[0m homologous genes found. Data-level neighbor search may yield noisy results.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">context_ind</span><span class="p">))))</span>      
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Found </span><span class="se">\033</span><span class="s2">[35m</span><span class="si">{}</span><span class="se">\033</span><span class="s2">[0m shared homologous genes between context and target dataset&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">context_ind</span><span class="p">))))</span>


        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Perform the data-level nearest neigbor search on the homologous gene set.&#39;</span><span class="p">)</span>    

        <span class="k">if</span> <span class="n">compute_log1p</span><span class="p">:</span>
            <span class="n">context_neigh</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataset_collection</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">context_dataset_name</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()[:,</span> <span class="n">context_ind</span><span class="p">])</span>
            <span class="n">target_neigh</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()[:,</span> <span class="n">target_ind</span><span class="p">])</span>

        <span class="k">else</span><span class="p">:</span>    
            <span class="n">context_neigh</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset_collection</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">context_dataset_name</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()[:,</span> <span class="n">context_ind</span><span class="p">]</span>
            <span class="n">target_neigh</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()[:,</span> <span class="n">target_ind</span><span class="p">]</span>

        <span class="n">neigh</span> <span class="o">=</span> <span class="n">NearestNeighbors</span><span class="p">(</span><span class="o">**</span><span class="n">nn_kwargs</span><span class="p">)</span>
        <span class="n">neigh</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">context_neigh</span><span class="p">)</span>

        <span class="n">_</span><span class="p">,</span> <span class="n">indices_whole</span> <span class="o">=</span> <span class="n">neigh</span><span class="o">.</span><span class="n">kneighbors</span><span class="p">(</span><span class="n">target_neigh</span><span class="p">)</span>
        <span class="n">indices_whole</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(</span><span class="n">indices_whole</span><span class="p">)</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;ind_neigh_nns&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">indices_whole</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">eval_nns_keys</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">eval_nns_keys</span> <span class="o">=</span> <span class="p">[</span><span class="n">cell_key</span><span class="p">]</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_labels_nns_hom_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">eval_nns_keys</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">dataset_collection</span><span class="p">[</span><span class="n">dataset_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done!</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">+</span><span class="s1">&#39;-&#39;</span><span class="o">*</span><span class="mi">90</span><span class="p">)</span></div>


<div class="viewcode-block" id="create_mdata.translate_gene_list">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.translate_gene_list">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">translate_gene_list</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">        </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translate gene symbols in var_names of a target AnnData to homologous context-species symbols.</span>

<span class="sd">        Will download a HOM_AllOrganism.rpt if not present if context-target species pair consits of </span>
<span class="sd">        human, mouse, rat or zebrafish. Will fallback to `map_homologs_silent` for unsupported species pairs.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : anndata.AnnData</span>
<span class="sd">            Target AnnData whose var_names will be translated.</span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - Prints a status message about which datasets are being translated.</span>
<span class="sd">        - Downloads and saves `HOM_AllOrganism.rpt` if not already present.</span>
<span class="sd">        - Reads the homology report into a DataFrame.</span>
<span class="sd">        - Filters the table to context and target species.</span>
<span class="sd">        - Computes a translated gene list via `get_key` or falls back to `map_homologs_silent` if species is not human, mouse, rat or zebrafish.</span>
<span class="sd">        - Sets `adata.var[&#39;var_names_transl&#39;]` to the mapped names.            </span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Translating homologous gene names between </span><span class="si">{}</span><span class="s1"> context and </span><span class="si">{}</span><span class="s1"> target dataset.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">context_dataset_name</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]))</span>
        <span class="n">gene_list</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span>
        <span class="n">NCBI_Taxon_ID</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;NCBI_Taxon_ID&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">context_NCBI_Taxon_ID</span> <span class="o">==</span> <span class="n">NCBI_Taxon_ID</span><span class="p">:</span>
            <span class="n">transl_gene_list</span> <span class="o">=</span> <span class="n">gene_list</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">context_NCBI_Taxon_ID</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">9606</span><span class="p">,</span> <span class="mi">10090</span><span class="p">,</span> <span class="mi">10116</span><span class="p">,</span> <span class="mi">7955</span><span class="p">)</span> <span class="ow">and</span> <span class="n">NCBI_Taxon_ID</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">9606</span><span class="p">,</span> <span class="mi">10090</span><span class="p">,</span> <span class="mi">10116</span><span class="p">,</span> <span class="mi">7955</span><span class="p">):</span> <span class="c1">#False:# s</span>
            <span class="n">out_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;data&quot;</span><span class="p">)</span>
            <span class="n">out_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">out_file</span> <span class="o">=</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="s2">&quot;HOM_AllOrganism.rpt&quot;</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">out_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Downloading gene translation dictionary.&quot;</span><span class="p">)</span>
                <span class="n">homology_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;https://www.informatics.jax.org/downloads/reports/HOM_AllOrganism.rpt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">homology_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene translation dictionary saved to data/HOM_AllOrganism.rpt&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_file</span><span class="p">))</span>

            <span class="k">else</span><span class="p">:</span>    
                <span class="n">homology_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">homology_contsp_df</span> <span class="o">=</span> <span class="n">homology_df</span><span class="p">[</span><span class="n">homology_df</span><span class="p">[</span><span class="s1">&#39;NCBI Taxon ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">context_NCBI_Taxon_ID</span><span class="p">]</span>
            <span class="n">homology_targsp_df</span> <span class="o">=</span> <span class="n">homology_df</span><span class="p">[</span><span class="n">homology_df</span><span class="p">[</span><span class="s1">&#39;NCBI Taxon ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">NCBI_Taxon_ID</span><span class="p">]</span>
            <span class="n">transl_gene_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">get_key</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">homology_targsp_df</span><span class="p">,</span> <span class="n">homology_contsp_df</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">transl_gene_list</span> <span class="o">=</span> <span class="n">map_homologs_silent</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">gene_list</span><span class="p">),</span> <span class="n">NCBI_Taxon_ID</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">context_NCBI_Taxon_ID</span><span class="p">)</span>


        <span class="n">num_hom_genes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">transl_gene_list</span> <span class="k">if</span> <span class="s1">&#39;non_hom&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gene</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Could map </span><span class="se">\033</span><span class="s1">[33m</span><span class="si">{}</span><span class="se">\033</span><span class="s1">[0m of </span><span class="si">{}</span><span class="s1"> target gene symbols to context species gene symbols&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num_hom_genes</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">transl_gene_list</span><span class="p">))))</span>    

        <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;var_names_transl&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">transl_gene_list</span>
        <span class="k">return</span> <span class="n">adata</span></div>



<div class="viewcode-block" id="create_mdata.filter_cells">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.filter_cells">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">filter_cells</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> 
            <span class="n">min_non_zero_genes</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> 
            <span class="n">min_cell_type_size</span><span class="p">:</span> <span class="nb">int</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">        </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Filter cells based on minimum non-zero gene fraction and cell‐type size.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : anndata.AnnData</span>
<span class="sd">            The annotated data matrix to filter.</span>
<span class="sd">        min_non_zero_genes : float</span>
<span class="sd">            Minimum fraction of genes that must have nonzero counts in a cell.</span>
<span class="sd">        min_cell_type_size : int</span>
<span class="sd">            Minimum number of cells required to retain any given cell‐type.</span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - Removes cells with fewer than `min_non_zero_genes * n_vars` detected genes.</span>
<span class="sd">        - If a cell‐type key is set in `adata.uns[&#39;metadata&#39;][&#39;cell_key&#39;]`, discards</span>
<span class="sd">        any cell‐types with fewer than `min_cell_type_size` cells.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">old_n_obs</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span>
        <span class="n">cell_key</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;cell_key&#39;</span><span class="p">]</span>        
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">n_vars</span><span class="o">*</span><span class="n">min_non_zero_genes</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">cell_key</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">cell_type_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_key</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">&gt;</span><span class="n">min_cell_type_size</span>
            <span class="n">cell_type_counts</span> <span class="o">=</span> <span class="n">cell_type_counts</span><span class="p">[</span><span class="n">cell_type_counts</span><span class="o">==</span><span class="kc">True</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_key</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cell_type_counts</span><span class="p">)]</span>    

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Filtering cells. Kept </span><span class="si">{}</span><span class="s1">, removed </span><span class="si">{}</span><span class="s1">.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">old_n_obs</span><span class="o">-</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">))))</span>
        <span class="k">return</span> <span class="n">adata</span>        </div>


<div class="viewcode-block" id="create_mdata.compute_library_prior_params">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.compute_library_prior_params">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">compute_library_prior_params</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">        </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Compute scVI library size prior parameters for each cell.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : anndata.AnnData</span>
<span class="sd">            Annotated data matrix with raw counts in `adata.X`.</span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - Within each batch (from `adata.uns[&#39;metadata&#39;][&#39;batch_key&#39;]`),</span>
<span class="sd">        calculates the mean and standard deviation of log-total counts.</span>
<span class="sd">        - Stores values in `adata.obs[&#39;library_log_mean&#39;]` and</span>
<span class="sd">        `adata.obs[&#39;library_log_std&#39;]` as float32 columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Compute prior parameters for the library encoder.&#39;</span><span class="p">)</span>
        <span class="n">batch_key</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;batch_key&#39;</span><span class="p">]</span> 

        <span class="n">library_log_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">library_log_std</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>  
        <span class="n">log_sum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">batch</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]):</span>
            <span class="n">ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">batch</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">library_log_mean</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span>  <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">log_sum</span><span class="p">[</span><span class="n">ind</span><span class="p">])</span>
            <span class="n">library_log_std</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">log_sum</span><span class="p">[</span><span class="n">ind</span><span class="p">])</span>   

        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;library_log_mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">library_log_mean</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span> 
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;library_log_std&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">library_log_std</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span> 
        <span class="k">return</span> <span class="n">adata</span></div>



<div class="viewcode-block" id="create_mdata.pred_labels_nns_hom_genes">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.pred_labels_nns_hom_genes">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">pred_labels_nns_hom_genes</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> 
            <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> 
            <span class="n">context_label_keys</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">25</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">        </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Predicts target cell-type labels using data-level k-nearest neighbor search </span>
<span class="sd">        results over homologous genes shared with the context dataset.</span>
<span class="sd">        Additionaly calculates the uncertainty score that will be used by scSpecies to decide</span>
<span class="sd">        which cells are aligned during fine-tuning.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : anndata.AnnData</span>
<span class="sd">            Target dataset that contains the neighbor indices in `adata.obsm[&#39;ind_neigh_nns&#39;]`.</span>
<span class="sd">        context_label_keys : list of str</span>
<span class="sd">            Keys in the context dataset&#39;s `obs` corresponding to categorical labels to be transferred </span>
<span class="sd">            (e.g., cell-type, tissue-type). </span>
<span class="sd">        k : int</span>
<span class="sd">            Amount of neighbort to consider for majority voting</span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - For each key in `context_label_keys`, assigns:</span>
<span class="sd">            - `adata.obs[&#39;pred_nns_&lt;label_key&gt;&#39;]`: predicted label (most frequent among neighbors).</span>
<span class="sd">            - `adata.obs[&#39;top_percent_&lt;label_key&gt;&#39;]`: confidence score based on relative neighbor rank.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">context_adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset_collection</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">context_dataset_name</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">context_label_key</span> <span class="ow">in</span> <span class="n">context_label_keys</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Evaluating data level NNS and calculating cells with the highest agreement for context labels key </span><span class="si">{}</span><span class="s1">.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">context_label_key</span><span class="p">))</span>

            <span class="n">ind_neigh</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;ind_neigh_nns&#39;</span><span class="p">]</span>  
            <span class="n">ind_neigh</span> <span class="o">=</span> <span class="n">ind_neigh</span><span class="p">[:,:</span><span class="n">k</span><span class="p">]</span>
            <span class="n">context_labels</span> <span class="o">=</span> <span class="n">context_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">context_label_key</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

            <span class="n">label_counts</span> <span class="o">=</span> <span class="p">[</span><span class="nb">dict</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">context_labels</span><span class="p">[</span><span class="n">ind_neigh</span><span class="p">[</span><span class="n">i</span><span class="p">]]))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">ind_neigh</span><span class="p">)[</span><span class="mi">0</span><span class="p">])]</span>
            <span class="n">top_dict</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="n">label_counts</span> <span class="o">=</span> <span class="p">[</span><span class="nb">max</span><span class="p">(</span><span class="n">label_counts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">ind_neigh</span><span class="p">)[</span><span class="mi">0</span><span class="p">])]</span>

            <span class="n">top_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">context_labels</span><span class="p">)}</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">label_counts</span><span class="p">)):</span>
                <span class="n">top_dict</span><span class="p">[</span><span class="n">label_counts</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">label_counts</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>

            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">top_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">top_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">top_dict</span><span class="p">[</span><span class="n">key</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">num_samples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">top_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>
                <span class="n">top_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">top_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="n">num_samples</span><span class="p">,)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">top_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]))]</span> 

            <span class="n">label_counts</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">top_dict</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">])</span> 

            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;top_percent_&#39;</span><span class="o">+</span><span class="n">context_label_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">label_counts</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">label_counts</span><span class="p">))])</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pred_nns_&#39;</span><span class="o">+</span><span class="n">context_label_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">label_counts</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">label_counts</span><span class="p">))])</span>
        <span class="k">return</span> <span class="n">adata</span>    </div>


<div class="viewcode-block" id="create_mdata.encode_batch_labels">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.encode_batch_labels">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">encode_batch_labels</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> 
            <span class="n">min_batch_size</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>    
<span class="w">         </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        One‐hot encode experimental batch labels, excluding small batches.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : anndata.AnnData</span>
<span class="sd">            Annotated data matrix with batch labels in `adata.obs[...]`.</span>
<span class="sd">        min_batch_size : int</span>
<span class="sd">            Smallest batch size to keep; batches with fewer cells are removed, must be &gt;= 0.</span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - Drops any batch categories with fewer than `min_batch_size` cells.</span>
<span class="sd">        - Fits a OneHotEncoder to remaining batch labels.</span>
<span class="sd">        - Saves the encoded batch matrix to `adata.obsm[&#39;batch_label_enc&#39;]`.</span>
<span class="sd">        - Builds `adata.uns[batch_dict]`, mapping each cell‐type (and &#39;unknown&#39;)</span>
<span class="sd">        to batch labels in which they have samples.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">min_batch_size</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">min_batch_size</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="n">batch_key</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;batch_key&#39;</span><span class="p">]</span> 
        <span class="n">cell_key</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;cell_key&#39;</span><span class="p">]</span> 
        <span class="n">name</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> 

        <span class="n">batch_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
        <span class="n">to_remove</span>   <span class="o">=</span> <span class="n">batch_counts</span><span class="p">[</span><span class="n">batch_counts</span> <span class="o">&lt;</span> <span class="n">min_batch_size</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
        <span class="n">adata</span>       <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="o">~</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">to_remove</span><span class="p">)]</span>
        <span class="n">batch_labels</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Registering experimental batches for the </span><span class="si">{}</span><span class="s1"> dataset. Kept </span><span class="si">{}</span><span class="s1">, removed </span><span class="si">{}</span><span class="s1">.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">batch_labels</span><span class="p">))),</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">batch_counts</span><span class="p">))))</span>

        <span class="n">enc</span> <span class="o">=</span> <span class="n">OneHotEncoder</span><span class="p">()</span>
        <span class="n">enc</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">batch_labels</span><span class="p">)</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;batch_label_enc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enc</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">batch_labels</span><span class="p">)</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span> 

        <span class="k">if</span> <span class="n">cell_key</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">batch_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;unknown&#39;</span><span class="p">:</span> <span class="n">enc</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">batch_labels</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cell_types</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
            <span class="n">batch_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">3</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cell_types</span><span class="p">}</span>
            <span class="n">batch_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span> <span class="p">:</span> <span class="n">batch_dict</span><span class="p">[</span><span class="n">c</span><span class="p">][</span><span class="n">batch_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cell_types</span><span class="p">}</span>
            <span class="n">batch_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span> <span class="p">:</span> <span class="n">enc</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">batch_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>  <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cell_types</span><span class="p">}</span>
            <span class="n">batch_dict</span><span class="p">[</span><span class="s1">&#39;unknown&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enc</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">batch_labels</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;batch_dict&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">batch_dict</span>

        <span class="k">return</span> <span class="n">adata</span></div>



<div class="viewcode-block" id="create_mdata.subset_to_highly_variable_genes">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.subset_to_highly_variable_genes">[docs]</a>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">subset_to_highly_variable_genes</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>                                       
            <span class="n">n_top_genes</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Subset dataset to the top highly variable genes using the Seurat method.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : anndata.AnnData</span>
<span class="sd">            Annotated data matrix to subset.</span>
<span class="sd">        n_top_genes : int</span>
<span class="sd">            Number of top highly variable genes to select.</span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - Subsets `adata` to the top `n_top_genes` hvg genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>        
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Subsetting the </span><span class="si">{}</span><span class="s1"> dataset to the </span><span class="si">{}</span><span class="s1"> most highly variable genes using seurat.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="nb">str</span><span class="p">(</span><span class="n">n_top_genes</span><span class="p">)))</span>

        <span class="n">batch_key</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;metadata&#39;</span><span class="p">][</span><span class="s1">&#39;batch_key&#39;</span><span class="p">]</span> 

        <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;raw_counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> 
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">,</span>
            <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">,</span>
            <span class="n">n_top_genes</span><span class="o">=</span><span class="n">n_top_genes</span><span class="p">,</span>
            <span class="n">subset</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">flavor</span><span class="o">=</span><span class="s1">&#39;seurat_v3&#39;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;raw_counts&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="c1">#.toarray()</span>
        <span class="k">del</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;raw_counts&#39;</span><span class="p">]</span>        
        
        <span class="c1">#del adata.var[&#39;highly_variable_intersection&#39;]</span>
        <span class="c1">#del adata.var[&#39;dispersions_norm&#39;]</span>
        <span class="c1">#del adata.var[&#39;dispersions&#39;]</span>
        <span class="c1">#del adata.var[&#39;means&#39;]</span>
        <span class="c1">#del adata.var[&#39;highly_variable&#39;]</span>
        <span class="c1">#del adata.var[&#39;highly_variable_nbatches&#39;]       </span>
        <span class="c1">#del adata.uns[&#39;log1p&#39;]</span>
        <span class="c1">#del adata.uns[&#39;hvg&#39;] </span>

        <span class="k">return</span> <span class="n">adata</span></div>


<div class="viewcode-block" id="create_mdata.return_mdata">
<a class="viewcode-back" href="../../scspecies.html#scspecies.preprocessing.create_mdata.return_mdata">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">return_mdata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> 
            <span class="n">return_mdata</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> 
            <span class="n">save</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> 
            <span class="n">save_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;data&quot;</span><span class="p">),</span> 
            <span class="n">save_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;mudata&#39;</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">mu</span><span class="o">.</span><span class="n">MuData</span><span class="p">:</span>
<span class="w">        </span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Optionally save and/or return the assembled MuData object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        return_mdata : bool, optional</span>
<span class="sd">            If True, return the MuData object at the end (default True).</span>
<span class="sd">        save : bool, optional</span>
<span class="sd">            If True, write the MuData object to disk (default True).</span>
<span class="sd">        save_path : pathlib.Path, optional</span>
<span class="sd">            Directory in which to save the file; created if missing (default Path(&quot;data&quot;)).</span>
<span class="sd">        save_name : str, optional</span>
<span class="sd">            Filename stem for the .h5mu file; &#39;.h5mu&#39; is appended (default &#39;mudata&#39;).</span>

<span class="sd">        Effects</span>
<span class="sd">        -------</span>
<span class="sd">        - If `save` is True:</span>
<span class="sd">            - Ensures that `save_path` exists, creating it if necessary.</span>
<span class="sd">            - Writes the MuData assembled from `self.dataset_collection` to</span>
<span class="sd">            `save_path/&lt;save_name&gt;.h5mu`.</span>
<span class="sd">            - Prints messages about directory creation and file saving.</span>
<span class="sd">        - If `return_mdata` is True:</span>
<span class="sd">            - Returns the MuData object constructed from `self.dataset_collection`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">save</span><span class="p">:</span>
            <span class="n">save_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">save_path</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">save_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="n">save_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Created directory &#39;</span><span class="si">{</span><span class="n">save_path</span><span class="si">}</span><span class="s2">&#39;.&quot;</span><span class="p">)</span>

            <span class="n">mdata</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">MuData</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dataset_collection</span><span class="p">)</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">save_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">save_name</span><span class="si">}</span><span class="s2">.h5mu&quot;</span> 

            <span class="n">mdata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">file_path</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved mdata to </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">return_mdata</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">mdata</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Clemens Schächter.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>