

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>The scSpecies workflow &mdash; scspecies Documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/jupyter-sphinx.css?v=572af1d6" />
      <link rel="stylesheet" type="text/css" href="../_static/thebelab.css" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=01f34227"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/thebelab-helper.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tips to fit scSpecies" href="../tips.html" />
    <link rel="prev" title="Set up datasets for scSpecies" href="data_preprocessing.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            scspecies
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Navigation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api/api.html">API Reference</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="data_preprocessing.html">Set up datasets for scSpecies</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">The scSpecies workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#1)-Context-and-target-dataset-alignment">1) Context and target dataset alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#2)-Information-transfer-via-similarity-scores">2) Information transfer via similarity scores</a></li>
<li class="toctree-l3"><a class="reference internal" href="#3)-Differential-gene-expression-analysis">3) Differential gene expression analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#4)-Creating-a-cell-atlas">4) Creating a cell atlas</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tips.html">Tips to fit scSpecies</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scspecies</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">The scSpecies workflow</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/aligning_datasets.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="The-scSpecies-workflow">
<h1>The scSpecies workflow<a class="headerlink" href="#The-scSpecies-workflow" title="Link to this heading">ÔÉÅ</a></h1>
<p>This tutorial demonstrates how to apply the scSpecies workflow to align three scRNA-seq datasets (mice, humans and hamsters).</p>
<p>We start by loading the preprocesed <code class="docutils literal notranslate"><span class="pre">.h5mu</span></code> file saved within the <code class="docutils literal notranslate"><span class="pre">data_prepocessing.ipynb</span></code> notebook.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">muon</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mu</span>

<span class="n">mu</span><span class="o">.</span><span class="n">set_options</span><span class="p">(</span><span class="n">pull_on_update</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1">#os.environ[&quot;TORCHDYNAMO_DISABLE&quot;] = &quot;1&quot;</span>

<span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">autoreload</span> 2

<span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\\</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;/&#39;</span>
<span class="n">mdata</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">read_h5mu</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s1">&#39;data/liver_atlas.h5mu&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload
</pre></div></div>
</div>
<p>Before we pre-train scSpecies, we plot a UMAP representation of the unaligned mouse/human dataset pair on the data-level.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">sparse</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">return_palette</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.models</span><span class="w"> </span><span class="kn">import</span> <span class="n">neighbors_workaround</span>

<span class="c1"># Subsetting to homologous genes</span>
<span class="n">_</span><span class="p">,</span> <span class="n">hom_ind_mouse</span><span class="p">,</span> <span class="n">hom_ind_human</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;var_names_transl&#39;</span><span class="p">],</span> <span class="n">return_indices</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">adata_concat</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span>
    <span class="n">X</span><span class="o">=</span><span class="n">sparse</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">][:,</span> <span class="n">hom_ind_mouse</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">][:,</span> <span class="n">hom_ind_human</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">])</span><span class="o">.</span><span class="n">toarray</span><span class="p">(),</span>
    <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">])</span>
<span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># Color scheme for the liver cell dataset. Won&#39;t return nice results for other datasets.</span>
<span class="n">palette</span> <span class="o">=</span> <span class="n">return_palette</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">adata_concat</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type_fine</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata_concat</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">)</span>

<span class="c1"># sc.pp.neighbors might crash the kernel for M1/M2 chips and use a workaroung for this function</span>
<span class="c1">#sc.pp.neighbors(adata_concat, use_rep=&#39;X_pca&#39;)</span>
<span class="n">neighbors_workaround</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type_fine&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_3_0.png" src="../_images/tutorials_aligning_datasets_3_0.png" />
</div>
</div>
<section id="1)-Context-and-target-dataset-alignment">
<h2>1) Context and target dataset alignment<a class="headerlink" href="#1)-Context-and-target-dataset-alignment" title="Link to this heading">ÔÉÅ</a></h2>
<div class="line-block">
<div class="line">We start by aligning the mouse and human scRNA-seq datasets using scSpecies.</div>
<div class="line">We use the mouse dataset as context for the human target dataset.</div>
<div class="line">We define the corresponding context and target scVI models by instantiating the <code class="docutils literal notranslate"><span class="pre">scSpecies</span></code> class.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.models</span><span class="w"> </span><span class="kn">import</span> <span class="n">scSpecies</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>

<span class="n">device</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;mps&quot;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">mps</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="ow">and</span> <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">mps</span><span class="o">.</span><span class="n">is_built</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;cuda&quot;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span><span class="p">)</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">scSpecies</span><span class="p">(</span><span class="n">device</span><span class="p">,</span>
                <span class="n">mdata</span><span class="p">,</span>
                <span class="n">path</span><span class="p">,</span>
                <span class="n">context_key</span> <span class="o">=</span> <span class="s1">&#39;mouse&#39;</span><span class="p">,</span>
                <span class="n">target_key</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">,</span>
                <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span>
                <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initializing context scVI model.
Initializing target scVI model.
</pre></div></div>
</div>
<div class="line-block">
<div class="line">First, we pre-train context scVI model with <code class="docutils literal notranslate"><span class="pre">train_context</span></code> for 30 epochs and save the model parameters to path.</div>
<div class="line">Afterwards, we calculate latent and intermediate representations with <code class="docutils literal notranslate"><span class="pre">get_representation</span></code> an save them in the context modality in the <code class="docutils literal notranslate"><span class="pre">.obsm</span></code> layer.</div>
<div class="line">Intermediate representations are also saved and later used during fine-tuning for alignment.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">train_context</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="n">save_key</span><span class="o">=</span><span class="s1">&#39;_mouse&#39;</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">get_representation</span><span class="p">(</span><span class="n">eval_model</span><span class="o">=</span><span class="s1">&#39;context&#39;</span><span class="p">,</span> <span class="n">save_intermediate</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">save_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Pretraining on the context dataset for 30 epochs (= 8820 iterations).
<span class="ansi-magenta-intense-fg">Progress</span>: 99.9%<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">ETA</span>: 0:00:00<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">Epoch</span>: 30<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">Iteration</span>: 8810<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">ms/Iteration</span>: 20.59<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">nELBO</span>: 1465.4 (<span class="ansi-red-intense-fg">-0.495</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">nlog_likeli</span>: 1447.1 (<span class="ansi-red-intense-fg">-0.450</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">KL-Div z</span>: 15.270 (<span class="ansi-red-intense-fg">-0.037</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">KL-Div l</span>: 3.0551 (<span class="ansi-red-intense-fg">-0.008</span>).           Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/config_dict.pkl
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/context_config__mouse.pkl
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/context_optimizer__mouse.opt
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/target_encoder_inner__mouse.pth.
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/context_encoder_outer__mouse.pth.
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/context_decoder__mouse.pth.
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/context_lib_encoder__mouse.pth.
Calculate latent variables. Step 167/295
</pre></div></div>
</div>
<p>We check that the context scVI model has separated latent cell clusters by visualizing the latent representation with UMAP.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#sc.pp.neighbors(model.mdata[&#39;mouse&#39;], use_rep=&#39;z_mu&#39;)</span>
<span class="n">neighbors_workaround</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">],</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;z_mu&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_9_0.png" src="../_images/tutorials_aligning_datasets_9_0.png" />
</div>
</div>
<p>Next we fine-tune by training the target scVI model and aligning the human dataset with this latent representations.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">train_target</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="n">save_key</span><span class="o">=</span><span class="s1">&#39;_human&#39;</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">get_representation</span><span class="p">(</span><span class="n">eval_model</span><span class="o">=</span><span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="n">save_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Training on the target dataset for 30 epochs (= 8190 iterations).
<span class="ansi-magenta-intense-fg">Progress</span>: 99.9%<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">ETA</span>: 0:00:00<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">Epoch</span>: 30<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">Iteration</span>: 8179<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">ms/Iteration</span>: 42.34<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">nELBO</span>: 1903.5 (<span class="ansi-green-intense-fg">+10.46</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">nlog_likeli</span>: 1384.8 (<span class="ansi-green-intense-fg">+0.451</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">KL-Div z</span>: 13.827 (<span class="ansi-red-intense-fg">-0.002</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">KL-Div l</span>: 2.8157 (<span class="ansi-red-intense-fg">-0.012</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">Align-Term</span>: 502.01 (<span class="ansi-green-intense-fg">+10.02</span>).           Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/config_dict.pkl
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/target_config__human.pkl
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/target_optimizer__human.opt
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/target_encoder_inner__human.pth.
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/target_encoder_outer__human.pth.
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/target_decoder__human.pth.
Saved /Users/cschaech/Desktop/scpecies_package/scspecies/tutorials/params/target_lib_encoder__human.pth.
Calculate latent variables. Step 181/274
</pre></div></div>
</div>
<p>After fine-tuning, we can visualize the aligned latent representation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_concat</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;lat_rep&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;z_mu&#39;</span><span class="p">],</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;z_mu&#39;</span><span class="p">]])</span>

<span class="c1">#sc.pp.neighbors(adata_concat, use_rep=&#39;lat_rep&#39;)</span>
<span class="n">neighbors_workaround</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;lat_rep&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type_fine&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_13_0.png" src="../_images/tutorials_aligning_datasets_13_0.png" />
</div>
</div>
</section>
<section id="2)-Information-transfer-via-similarity-scores">
<h2>2) Information transfer via similarity scores<a class="headerlink" href="#2)-Information-transfer-via-similarity-scores" title="Link to this heading">ÔÉÅ</a></h2>
<p>We continue by evaluating the likelihood based similarity measure between target and context cell types with <code class="docutils literal notranslate"><span class="pre">return_similarity_df_prot</span></code>.</p>
<div class="line-block">
<div class="line">When <code class="docutils literal notranslate"><span class="pre">scale=min_max</span></code> or <code class="docutils literal notranslate"><span class="pre">scale=max</span></code> the scores are scaled such that the values can be interpreted as 1 = most similar. Without scaling the highest values (closest to zero) represent high similarity. The output dataframe contains target cell types in <code class="docutils literal notranslate"><span class="pre">df.index</span></code> and context cell types in <code class="docutils literal notranslate"><span class="pre">df.columns</span></code>.</div>
<div class="line">The values should be interpreted row-wise and contain the similarity of a target cell type with the corresponding context cell types.</div>
</div>
<p>We can use these similarity scores to match cell type annotation between datasets. To reduce computational demand reduce the sample size via <code class="docutils literal notranslate"><span class="pre">max_sample_targ</span></code> and <code class="docutils literal notranslate"><span class="pre">max_sample_cont</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_prototype_sim_heatmap</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">return_similarity_df</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>

<span class="n">plot_prototype_sim_heatmap</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
                     B Cells   Basophils  CD8 Eff. Memory T  \
B Cells            -4.667412 -351.202118        -187.310165
Basophils        -342.285217   -6.553959        -323.325165
CD4+ KLRB1 Th    -233.442261 -248.224075         -78.080002
Central Vein ECs -200.179871 -881.981323        -306.111633
Cholangiocytes   -219.242920 -664.898621        -577.777588

                  Capsule Fibroblasts  Central Vein ECs  Cholangiocytes  \
B Cells                   -196.515381       -311.335846     -188.958267
Basophils                 -243.916443       -286.631592     -258.919800
CD4+ KLRB1 Th             -135.655609       -246.467194     -215.919586
Central Vein ECs          -541.279541        -42.993996     -809.694702
Cholangiocytes            -245.021790       -393.667328      -11.980484

                  Cytotoxic CD8+  Fibroblast 1  Fibroblast 2  Hepatocytes  \
B Cells              -407.455353   -232.646133   -257.090210  -194.491119
Basophils            -307.394958   -260.239685   -267.619141  -273.679291
CD4+ KLRB1 Th         -49.746071   -223.400787   -227.490250  -263.032715
Central Vein ECs     -169.014023   -479.993774   -397.975464  -724.254578
Cholangiocytes       -319.386292   -563.276184   -468.055328  -528.212524

                  ...  Portal Vein ECs  Regulatory T  Stellate Cells  \
B Cells           ...      -208.995758   -155.023468     -186.692200
Basophils         ...      -280.190247   -309.364716     -279.297119
CD4+ KLRB1 Th     ...       -96.285042    -47.582340     -200.683044
Central Vein ECs  ...       -99.734467   -383.681885     -300.876831
Cholangiocytes    ...      -413.301270   -658.578491     -450.541626

                        Th 1       Th 17  Trans. Monocytes  \
B Cells          -474.892456 -191.938156       -287.087402
Basophils        -324.944702 -284.656860       -387.242432
CD4+ KLRB1 Th     -17.696383  -21.042255       -361.853027
Central Vein ECs -376.703979 -311.065033      -1000.916504
Cholangiocytes   -390.991882 -447.759644       -945.070984

                  Trans. Monocytes 2      cDCs 1      cDCs 2        pDCs
B Cells                  -271.993195 -260.032532 -185.189636 -183.290771
Basophils                -376.799103 -438.412292 -342.855072 -364.234924
CD4+ KLRB1 Th            -346.482056 -393.770264 -297.487732 -367.613434
Central Vein ECs         -949.445557 -814.816528 -914.407349 -516.116943
Cholangiocytes           -694.829834 -527.614502 -916.734131 -520.529663

[5 rows x 36 columns]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_15_1.png" src="../_images/tutorials_aligning_datasets_15_1.png" />
</div>
</div>
<p>We can use the similarity measure to infer target label annotation from context cells.</p>
<div class="line-block">
<div class="line">We draw a random target cell index and calculate similarity scores for all context cells with <code class="docutils literal notranslate"><span class="pre">transfer_info</span></code>.</div>
<div class="line">We transfer the context <code class="docutils literal notranslate"><span class="pre">.obs</span></code> labels specified in <code class="docutils literal notranslate"><span class="pre">context_obs_transfer</span></code>.</div>
<div class="line">The <code class="docutils literal notranslate"><span class="pre">df_neigbor</span></code> dataframe contains all context cells with corresponding indices sorted by similarity.</div>
<div class="line">The labels can be transferred via majority vote based on the occurence in the top-k most similar cells.</div>
</div>
<div class="line-block">
<div class="line">We can plot the similarity scores for the whole dataset using <code class="docutils literal notranslate"><span class="pre">plot_similarity</span></code>.</div>
<div class="line">For cell type label coloring refer to the aligned latent space visualization. Green colors indicate the similar cells while the target cell is marked by a black cross.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_similarity</span>

<span class="n">human_cell_types</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">]</span>
<span class="n">mouse_cell_types</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">mdata</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">]</span>
<span class="n">common_cell_types</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">human_cell_types</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">mouse_cell_types</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

<span class="n">human_inds</span> <span class="o">=</span> <span class="n">human_cell_types</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_cell_types</span><span class="p">)</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">human_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">human_inds</span><span class="p">)</span>

<span class="n">context_obs_transfer</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cell_type_coarse&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type_fine&#39;</span><span class="p">]</span>
<span class="n">df_neigbor</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">transfer_info</span><span class="p">(</span><span class="n">human_ind</span><span class="p">,</span> <span class="n">context_obs_transfer</span><span class="p">)</span>

<span class="c1">#print(&#39;Index of target human cell: {}, Information: {}.&#39;.format(str(human_ind), &#39;, &#39;.join([obs_name+&#39;: &#39;+label for label, obs_name in zip(model.mdata[&#39;human&#39;].obs[context_obs_transfer].iloc[human_ind].values[0], context_obs_transfer)])))</span>
<span class="c1">#print(df_neigbor.head())</span>

<span class="n">plot_similarity</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">,</span> <span class="n">df_neigbor</span><span class="p">,</span> <span class="n">human_ind</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_17_0.png" src="../_images/tutorials_aligning_datasets_17_0.png" />
</div>
</div>
<div class="line-block">
<div class="line">Next, we perform this process for the whole dataset using <code class="docutils literal notranslate"><span class="pre">label_transfer</span></code>.</div>
<div class="line">This transfers the specified context labels for the whole dataset and saves the predictions to the <code class="docutils literal notranslate"><span class="pre">.obs</span></code> layer of the target dataset.</div>
<div class="line">Afterwards, we can compute a normalized confusion matrix (%) and the total balanced accuracy for label transfer with <code class="docutils literal notranslate"><span class="pre">ret_pred_df</span></code>.</div>
<div class="line">We compare this against the results of the data-level neighbor search.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">context_obs_transfer</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cell_type_coarse&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type_fine&#39;</span><span class="p">]</span>
<span class="n">model</span><span class="o">.</span><span class="n">label_transfer</span><span class="p">(</span><span class="n">context_obs_transfer</span><span class="p">)</span>
<span class="c1">#print(model.mdata[&#39;human&#39;].obs[[&#39;pred_sim_cell_type_coarse&#39;, &#39;cell_type_coarse&#39;, &#39;pred_sim_cell_type_fine&#39;, &#39;cell_type_fine&#39;]].head())</span>

<span class="n">df_nns</span><span class="p">,</span> <span class="n">bas_nns</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">ret_pred_df</span><span class="p">(</span><span class="n">pred_key</span><span class="o">=</span><span class="s1">&#39;pred_nns_cell_type_fine&#39;</span><span class="p">,</span> <span class="n">target_label_key</span><span class="o">=</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">,</span> <span class="n">context_label_key</span><span class="o">=</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">)</span>
<span class="n">df_sim</span><span class="p">,</span> <span class="n">bas_sim</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">ret_pred_df</span><span class="p">(</span><span class="n">pred_key</span><span class="o">=</span><span class="s1">&#39;pred_sim_cell_type_fine&#39;</span><span class="p">,</span> <span class="n">target_label_key</span><span class="o">=</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">,</span> <span class="n">context_label_key</span><span class="o">=</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">)</span>

<span class="c1">#print(df_sim.head())</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data-level k=25 nearest neighbor search --&gt; Balanced accuracy: </span><span class="si">{}</span><span class="s1">%&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">bas_nns</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span><span class="mi">2</span><span class="p">)))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Label tarnsfer using similarity measure --&gt; Balanced accuracy: </span><span class="si">{}</span><span class="s1">%&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">bas_sim</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span><span class="mi">2</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Pre-computing latent space NNS with 250 neighbors using the euclidean distance.
Calculate similarity metric. Step 271/274.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/cschaech/Desktop/package_test/.venv/lib/python3.11/site-packages/sklearn/metrics/_classification.py:2480: UserWarning: y_pred contains classes not in y_true
  warnings.warn(&#34;y_pred contains classes not in y_true&#34;)
</pre></div></div>
</div>
<div class="line-block">
<div class="line">We compare the results visually by generateing a bar plot that indicates label transfer improvement over the data level NNS.</div>
<div class="line">For each target cell type the data NNS accuracy is displayed in the top row and the similarity based label transfer accuracy in the bottom row.</div>
<div class="line">The improvement in % is displayed at the right.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">label_transfer_acc</span>

<span class="n">label_transfer_acc</span><span class="p">(</span><span class="n">df_nns</span><span class="p">,</span> <span class="n">df_sim</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_21_0.png" src="../_images/tutorials_aligning_datasets_21_0.png" />
</div>
</div>
</section>
<section id="3)-Differential-gene-expression-analysis">
<h2>3) Differential gene expression analysis<a class="headerlink" href="#3)-Differential-gene-expression-analysis" title="Link to this heading">ÔÉÅ</a></h2>
<p>The difference in modeled gene expression can be analyzed by comparing the log2-fold change in normalized gene expression with <code class="docutils literal notranslate"><span class="pre">compute_logfold_change</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lfc_dict</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">compute_logfold_change</span><span class="p">(</span><span class="n">lfc_delta</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The output is a dictionary with cell type wise data frames containing logfoldchange values and other information. The dataframes contain the homologous traget gene symbols in their index and their chosen labels as columns.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">rho_median_context</span></code> : Contains median context normalized gene expression,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mu_median_context</span></code> : Contains median context expected value gene expression,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rho_median_target</span></code> : Contains median target normalized gene expression,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mu_median_target</span></code> : Contains median target expected value gene expression,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lfc</span></code> : Contains mMedian Log2 fold-change of the relative expression parameter rho,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">p</span></code> : Probability of Log2 fold-change values greater than <code class="docutils literal notranslate"><span class="pre">lfc_delta</span></code>,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lfc_rand</span></code> : Contains median Log2 fold-change of the relative expression parameter rho on permuted data,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">p_rand</span></code> : Probability of Log2 fold-change values greater than <code class="docutils literal notranslate"><span class="pre">lfc_delta</span></code> on permuted data.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Results for&#39;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">lfc_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">lfc_dict</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">lfc_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results for B Cells
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rho_median_context</th>
      <th>mu_median_context</th>
      <th>rho_median_target</th>
      <th>mu_median_target</th>
      <th>lfc</th>
      <th>p</th>
      <th>lfc_rand</th>
      <th>p_rand</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>C12orf75</th>
      <td>1.022538e-07</td>
      <td>0.000536</td>
      <td>0.000131</td>
      <td>0.190623</td>
      <td>6.909566</td>
      <td>1.000000</td>
      <td>-0.712836</td>
      <td>0.129802</td>
    </tr>
    <tr>
      <th>C1orf21</th>
      <td>2.473843e-06</td>
      <td>0.012912</td>
      <td>0.000008</td>
      <td>0.011527</td>
      <td>1.352708</td>
      <td>0.974835</td>
      <td>3.159477</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>C19orf33</th>
      <td>7.167783e-06</td>
      <td>0.037320</td>
      <td>0.000002</td>
      <td>0.002491</td>
      <td>-1.584132</td>
      <td>0.999980</td>
      <td>-3.594840</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>KIAA0513</th>
      <td>1.270223e-06</td>
      <td>0.006634</td>
      <td>0.000007</td>
      <td>0.010222</td>
      <td>1.808130</td>
      <td>1.000000</td>
      <td>-10.413317</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>C15orf48</th>
      <td>1.599265e-07</td>
      <td>0.000835</td>
      <td>0.000003</td>
      <td>0.005040</td>
      <td>1.932442</td>
      <td>1.000000</td>
      <td>-0.790993</td>
      <td>0.229895</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We can visualize the results with <code class="docutils literal notranslate"><span class="pre">plot_lfc</span></code> per cell type.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_lfc</span>

<span class="n">plot_lfc</span><span class="p">(</span><span class="n">lfc_dict</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_27_0.png" src="../_images/tutorials_aligning_datasets_27_0.png" />
</div>
</div>
<p>Lets compare the results with a DFG Analysis at the data level. For this we generate homologous cell samples from the latent space.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_lfc_comparison</span>

<span class="n">target_rho_dict</span><span class="p">,</span> <span class="n">context_rho_dict</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">generate_homologous_samples</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>
<span class="n">plot_lfc_comparison</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">lfc_dict</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line">We can calculate activity scores for pathways with a downloaded <code class="docutils literal notranslate"><span class="pre">.gmt</span></code> file.</div>
<div class="line">For demonstartive purposes we provide a manual pathway dictionary.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_and_filter_pathways</span>

<span class="n">plot_cell_type</span> <span class="o">=</span> <span class="s1">&#39;B Cells&#39;</span>

<span class="n">adata_h</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">target_rho_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">target_rho_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
<span class="n">adata_h</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">][:,</span> <span class="n">model</span><span class="o">.</span><span class="n">target_config</span><span class="p">[</span><span class="s1">&#39;homologous_genes&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">var_names</span>
<span class="n">adata_h</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concat</span><span class="p">([[</span><span class="n">key</span><span class="p">]</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">target_rho_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">target_rho_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
<span class="n">adata_h</span><span class="o">.</span><span class="n">obs_names_make_unique</span><span class="p">()</span>

<span class="n">adata_m</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">context_rho_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">context_rho_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
<span class="n">adata_m</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">adata_h</span><span class="o">.</span><span class="n">var_names</span>
<span class="n">adata_m</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type_fine&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concat</span><span class="p">([[</span><span class="n">key</span><span class="p">]</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">context_rho_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">context_rho_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
<span class="n">adata_m</span><span class="o">.</span><span class="n">obs_names_make_unique</span><span class="p">()</span>

<span class="c1">#gene_sets_path = &#39;/Users/cschaech/Desktop/scSpecies/dataset/c2.all.v2024.1.Hs.symbols.gmt&#39;</span>
<span class="c1">#pathways = load_and_filter_pathways(gene_sets_path, adata_h)</span>

<span class="n">pathways</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ABBUD_LIF_SIGNALING_1_DN&#39;</span><span class="p">:</span>           <span class="p">[</span><span class="s1">&#39;LIMS1&#39;</span><span class="p">,</span><span class="s1">&#39;ITGA6&#39;</span><span class="p">,</span><span class="s1">&#39;ENPP2&#39;</span><span class="p">,</span><span class="s1">&#39;AHNAK&#39;</span><span class="p">,</span><span class="s1">&#39;ALCAM&#39;</span><span class="p">,</span><span class="s1">&#39;KLRB1&#39;</span><span class="p">,</span><span class="s1">&#39;HK2&#39;</span><span class="p">],</span>
            <span class="s1">&#39;HUMMERICH_MALIGNANT_SKIN_TUMOR_UP&#39;</span><span class="p">:</span>  <span class="p">[</span><span class="s1">&#39;S100A9&#39;</span><span class="p">,</span> <span class="s1">&#39;S100A8&#39;</span><span class="p">,</span> <span class="s1">&#39;LTF&#39;</span><span class="p">,</span> <span class="s1">&#39;CCND1&#39;</span><span class="p">,</span> <span class="s1">&#39;GSTO1&#39;</span><span class="p">,</span><span class="s1">&#39;HBA2&#39;</span><span class="p">,</span> <span class="s1">&#39;COL18A1&#39;</span><span class="p">,</span><span class="s1">&#39;SLPI&#39;</span><span class="p">,</span><span class="s1">&#39;ECM1&#39;</span><span class="p">],</span>
            <span class="s1">&#39;HILLION_HMGA1_TARGETS&#39;</span><span class="p">:</span>              <span class="p">[</span><span class="s1">&#39;CRIP2&#39;</span><span class="p">,</span> <span class="s1">&#39;EDNRB&#39;</span><span class="p">,</span><span class="s1">&#39;HSPD1&#39;</span><span class="p">,</span><span class="s1">&#39;INSR&#39;</span><span class="p">,</span><span class="s1">&#39;GPX3&#39;</span><span class="p">,</span><span class="s1">&#39;CXCR4&#39;</span><span class="p">,</span><span class="s1">&#39;PMP22&#39;</span><span class="p">,</span><span class="s1">&#39;TIMP2&#39;</span><span class="p">,</span><span class="s1">&#39;ID2&#39;</span><span class="p">,</span><span class="s1">&#39;ID3&#39;</span><span class="p">,</span><span class="s1">&#39;MGST1&#39;</span><span class="p">,</span><span class="s1">&#39;CD3G&#39;</span><span class="p">,</span><span class="s1">&#39;ID1&#39;</span><span class="p">,</span><span class="s1">&#39;HSPB1&#39;</span><span class="p">,</span><span class="s1">&#39;TFRC&#39;</span><span class="p">,</span><span class="s1">&#39;CLU&#39;</span><span class="p">]}</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">adata_m</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
    <span class="n">adata_h</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;species&quot;</span><span class="p">,</span>
    <span class="n">batch_categories</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mouse&quot;</span><span class="p">,</span> <span class="s2">&quot;human&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">pathways</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">pathway</span><span class="p">,</span> <span class="n">score_name</span><span class="o">=</span><span class="n">key</span><span class="p">)</span>

    <span class="n">adata_plot</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type_fine</span> <span class="o">==</span> <span class="n">plot_cell_type</span><span class="p">]</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">adata_plot</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span>
        <span class="n">x</span><span class="o">=</span><span class="s1">&#39;species&#39;</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="n">key</span><span class="p">,</span>
        <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;species&#39;</span><span class="p">,</span>
        <span class="n">palette</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;mouse&#39;</span><span class="p">:</span><span class="s1">&#39;C1&#39;</span><span class="p">,</span><span class="s1">&#39;human&#39;</span><span class="p">:</span><span class="s1">&#39;C0&#39;</span><span class="p">},</span>
        <span class="n">dodge</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">inner</span><span class="o">=</span><span class="s1">&#39;quartile&#39;</span><span class="p">,</span>
        <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
    <span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Pathway: </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Species&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Activity Score&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/cschaech/Desktop/package_test/.venv/lib/python3.11/site-packages/anndata/_core/anndata.py:1756: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
  utils.warn_names_duplicates(&#34;obs&#34;)
/Users/cschaech/Desktop/package_test/.venv/lib/python3.11/site-packages/anndata/_core/anndata.py:1756: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
  utils.warn_names_duplicates(&#34;obs&#34;)
/var/folders/bq/cdql_9s11db1zmbxmw_gz7vw0000gn/T/ipykernel_32274/4126954639.py:22: FutureWarning: Use anndata.concat instead of AnnData.concatenate, AnnData.concatenate is deprecated and will be removed in the future. See the tutorial for concat at: https://anndata.readthedocs.io/en/latest/concatenation.html
  adata = adata_m.concatenate(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_31_1.png" src="../_images/tutorials_aligning_datasets_31_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_31_2.png" src="../_images/tutorials_aligning_datasets_31_2.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_31_3.png" src="../_images/tutorials_aligning_datasets_31_3.png" />
</div>
</div>
<p>Next we use the differentally expressed genes to analyze pathways. As an example we can create ORA Barplots.</p>
<ul class="simple">
<li><p>Y-axis: Lists the top five enriched pathways whose member genes are over-represented among the differentially expressed genes.</p></li>
<li><p>X-axis (adj. P-value): Longer bars mean more significant enrichment.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">gseapy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">gp</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">textwrap</span>

<span class="n">cell_type</span> <span class="o">=</span> <span class="s1">&#39;B Cells&#39;</span>
<span class="n">ORA_LIBS</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KEGG_2021_Human&#39;</span><span class="p">,</span> <span class="s1">&#39;Reactome_Pathways_2024&#39;</span><span class="p">]</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">lfc_dict</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span>
<span class="n">degs</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;lfc&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;p&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.9</span><span class="p">)]</span>

<span class="n">enr</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">enrichr</span><span class="p">(</span>
    <span class="n">gene_list</span><span class="o">=</span><span class="n">degs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
    <span class="n">gene_sets</span><span class="o">=</span><span class="n">ORA_LIBS</span><span class="p">,</span>
    <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span>
    <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
<span class="n">top5</span> <span class="o">=</span> <span class="n">enr</span><span class="o">.</span><span class="n">results</span><span class="p">[[</span><span class="s1">&#39;Term&#39;</span><span class="p">,</span><span class="s1">&#39;Adjusted P-value&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>

<span class="n">wrapped</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">textwrap</span><span class="o">.</span><span class="n">wrap</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">30</span><span class="p">))</span>
<span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">top5</span><span class="p">[</span><span class="s1">&#39;Term&#39;</span><span class="p">]]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span>
    <span class="n">x</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">top5</span><span class="p">[</span><span class="s1">&#39;Adjusted P-value&#39;</span><span class="p">]),</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">wrapped</span><span class="p">,</span>
    <span class="n">hue</span>     <span class="o">=</span> <span class="n">wrapped</span><span class="p">,</span>
    <span class="n">palette</span> <span class="o">=</span> <span class="s1">&#39;Spectral&#39;</span><span class="p">,</span>
    <span class="n">dodge</span>   <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span>
<span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s1">: ORA Top 5 Pathways&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;-log10(adj. P-value)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="n">labelsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">labelsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_33_0.png" src="../_images/tutorials_aligning_datasets_33_0.png" />
</div>
</div>
<p>We can also plot the GSEA enrichment curves. We plt the pathway that is the most significant hit by FDR.</p>
<ul class="simple">
<li><p>X-axis (Rank positions): Genes sorted from highest to lowest log-fold change.</p></li>
<li><p>Running Enrichment Score: Curve rises when a pathway gene is encountered and falls otherwise; the peak ES indicates where pathway members cluster in the ranked list.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">gseapy.plot</span><span class="w"> </span><span class="kn">import</span> <span class="n">gseaplot</span>

<span class="n">GSEA_LIB</span> <span class="o">=</span> <span class="s1">&#39;KEGG_2021_Human&#39;</span>

<span class="n">pre_res</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">prerank</span><span class="p">(</span>
    <span class="n">rnk</span>            <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;lfc&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
    <span class="n">gene_sets</span>      <span class="o">=</span> <span class="n">GSEA_LIB</span><span class="p">,</span>
    <span class="n">processes</span>      <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
    <span class="n">permutation_num</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">outdir</span>         <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span>

<span class="n">res</span> <span class="o">=</span> <span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;FDR q-val&#39;</span><span class="p">)</span>
<span class="n">top_term</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;Term&#39;</span><span class="p">]</span>
<span class="n">rd</span> <span class="o">=</span> <span class="n">pre_res</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="n">top_term</span><span class="p">]</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">gseaplot</span><span class="p">(</span>
    <span class="n">term</span>        <span class="o">=</span> <span class="n">top_term</span><span class="p">,</span>
    <span class="n">hits</span>        <span class="o">=</span> <span class="n">rd</span><span class="p">[</span><span class="s1">&#39;hits&#39;</span><span class="p">],</span>
    <span class="n">nes</span>         <span class="o">=</span> <span class="n">rd</span><span class="p">[</span><span class="s1">&#39;nes&#39;</span><span class="p">],</span>
    <span class="n">pval</span>        <span class="o">=</span> <span class="n">rd</span><span class="p">[</span><span class="s1">&#39;pval&#39;</span><span class="p">],</span>
    <span class="n">fdr</span>         <span class="o">=</span> <span class="n">rd</span><span class="p">[</span><span class="s1">&#39;fdr&#39;</span><span class="p">],</span>
    <span class="n">RES</span>         <span class="o">=</span> <span class="n">rd</span><span class="p">[</span><span class="s1">&#39;RES&#39;</span><span class="p">],</span>
    <span class="n">rank_metric</span> <span class="o">=</span> <span class="n">pre_res</span><span class="o">.</span><span class="n">ranking</span><span class="p">,</span>
<span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/var/folders/bq/cdql_9s11db1zmbxmw_gz7vw0000gn/T/ipykernel_32274/596071678.py:5: DeprecationWarning: processes is deprecated; use threads
  pre_res = gp.prerank(
2025-05-28 01:47:50,359 [WARNING] Duplicated values found in preranked stats: 0.06% of genes
The order of those genes will be arbitrary, which may produce unexpected results.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_35_1.png" src="../_images/tutorials_aligning_datasets_35_1.png" />
</div>
</div>
</section>
<section id="4)-Creating-a-cell-atlas">
<h2>4) Creating a cell atlas<a class="headerlink" href="#4)-Creating-a-cell-atlas" title="Link to this heading">ÔÉÅ</a></h2>
<div class="line-block">
<div class="line">Finally, we can also create a multi-species cell atlas by aligning a hamster scVI archtecture to the same mouse context model.</div>
<div class="line">We instantiate a second model and load the encoder parameters of the context encoder.</div>
</div>
<p>For differential gene expression analysis the context decoder should be retrained.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_hamster</span> <span class="o">=</span> <span class="n">scSpecies</span><span class="p">(</span><span class="n">device</span><span class="p">,</span>
                <span class="n">mdata</span><span class="p">,</span>
                <span class="n">path</span><span class="p">,</span>
                <span class="n">context_key</span> <span class="o">=</span> <span class="s1">&#39;mouse&#39;</span><span class="p">,</span>
                <span class="n">target_key</span> <span class="o">=</span> <span class="s1">&#39;hamster&#39;</span><span class="p">,</span>
                <span class="p">)</span>

<span class="n">model_hamster</span><span class="o">.</span><span class="n">train_context</span><span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="n">train_decoder_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">save_key</span><span class="o">=</span><span class="s1">&#39;_hamster&#39;</span><span class="p">)</span>

<span class="n">model_hamster</span><span class="o">.</span><span class="n">train_target</span><span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="n">save_key</span><span class="o">=</span><span class="s1">&#39;_hamster&#39;</span><span class="p">)</span>
<span class="n">model_hamster</span><span class="o">.</span><span class="n">get_representation</span><span class="p">(</span><span class="n">eval_model</span><span class="o">=</span><span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="n">save_libsize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initializing context scVI self.
Initializing target scVI self.
Pretraining on the context dataset for 25 epochs (= 6225 iterations).
<span class="ansi-magenta-intense-fg">Progress</span>: 34.7%<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">ETA</span>: 0:01:21<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">Epoch</span>: 9<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">Iteration</span>: 2159<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-magenta-intense-fg">ms/Iteration</span>: 20.15<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">nELBO</span>: 1952.8 (<span class="ansi-green-intense-fg">+3.943</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">nlog_likeli</span>: 1937.2 (<span class="ansi-green-intense-fg">+5.284</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">KL-Div z</span>: 13.926 (<span class="ansi-red-intense-fg">-1.163</span>)<span class="ansi-cyan-intense-fg"> - </span><span class="ansi-yellow-fg">KL-Div l</span>: 1.6353 (<span class="ansi-red-intense-fg">-0.178</span>).
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">KeyboardInterrupt</span>                         Traceback (most recent call last)
<span class="ansi-cyan-fg">Cell</span><span class="ansi-cyan-fg"> </span><span class="ansi-green-fg">In[127]</span><span class="ansi-green-fg">, line 8</span>
<span class="ansi-green-fg">      1</span> model_hamster = scSpecies(device,
<span class="ansi-green-fg">      2</span>                 mdata,
<span class="ansi-green-fg">      3</span>                 path,
<span class="ansi-green-fg">      4</span>                 context_key = <span class="ansi-yellow-fg">&#39;</span><span class="ansi-yellow-fg">mouse</span><span class="ansi-yellow-fg">&#39;</span>,
<span class="ansi-green-fg">      5</span>                 target_key = <span class="ansi-yellow-fg">&#39;</span><span class="ansi-yellow-fg">hamster</span><span class="ansi-yellow-fg">&#39;</span>,
<span class="ansi-green-fg">      6</span>                 )
<span class="ansi-green-fg">----&gt; </span><span class="ansi-green-fg">8</span> <span class="ansi-yellow-bg">model_hamster</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">train_context</span><span class="ansi-yellow-bg">(</span><span class="ansi-green-fg ansi-yellow-bg">25</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">train_decoder_only</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg ansi-bold" style="color: rgb(0,135,0)">True</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">save_key</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-fg ansi-yellow-bg">&#39;</span><span class="ansi-yellow-fg ansi-yellow-bg">_hamster</span><span class="ansi-yellow-fg ansi-yellow-bg">&#39;</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">     10</span> model_hamster.train_target(<span class="ansi-green-fg">25</span>, save_key=<span class="ansi-yellow-fg">&#39;</span><span class="ansi-yellow-fg">_hamster</span><span class="ansi-yellow-fg">&#39;</span>)
<span class="ansi-green-fg">     11</span> model_hamster.get_representation(eval_model=<span class="ansi-yellow-fg">&#39;</span><span class="ansi-yellow-fg">target</span><span class="ansi-yellow-fg">&#39;</span>, save_libsize=<span class="ansi-bold" style="color: rgb(0,135,0)">True</span>)

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">~/Desktop/scpecies_package/scspecies/models.py:2032</span>, in <span class="ansi-cyan-fg">scSpecies.train_context</span><span class="ansi-blue-fg">(self, epochs, raise_beta, save_model, train_decoder_only, save_key)</span>
<span class="ansi-green-fg">   2029</span>     nelbo = nelbo + <span style="color: rgb(0,135,0)">self</span>.context_config[<span class="ansi-yellow-fg">&#39;</span><span class="ansi-yellow-fg">beta</span><span class="ansi-yellow-fg">&#39;</span>] * l_kl_div
<span class="ansi-green-fg">   2031</span> nelbo.backward()
<span class="ansi-green-fg">-&gt; </span><span class="ansi-green-fg">2032</span> <span class="ansi-yellow-bg" style="color: rgb(0,135,0)">self</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">context_optimizer</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">step</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">   2033</span> <span style="color: rgb(0,135,0)">self</span>.context_likeli_hist_dict.append(nlog_likeli.item())
<span class="ansi-green-fg">   2035</span> <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> <span style="color: rgb(0,135,0)">self</span>.config_dict[<span class="ansi-yellow-fg">&#39;</span><span class="ansi-yellow-fg">use_lib_enc</span><span class="ansi-yellow-fg">&#39;</span>]:

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">~/Desktop/package_test/.venv/lib/python3.11/site-packages/torch/optim/optimizer.py:485</span>, in <span class="ansi-cyan-fg">Optimizer.profile_hook_step.&lt;locals&gt;.wrapper</span><span class="ansi-blue-fg">(*args, **kwargs)</span>
<span class="ansi-green-fg">    480</span>         <span class="ansi-bold" style="color: rgb(0,135,0)">else</span>:
<span class="ansi-green-fg">    481</span>             <span class="ansi-bold" style="color: rgb(0,135,0)">raise</span> <span class="ansi-bold" style="color: rgb(215,95,95)">RuntimeError</span>(
<span class="ansi-green-fg">    482</span>                 <span class="ansi-yellow-fg">f</span><span class="ansi-yellow-fg">&#34;</span><span class="ansi-bold" style="color: rgb(175,95,135)">{</span>func<span class="ansi-bold" style="color: rgb(175,95,135)">}</span><span class="ansi-yellow-fg"> must return None or a tuple of (new_args, new_kwargs), but got </span><span class="ansi-bold" style="color: rgb(175,95,135)">{</span>result<span class="ansi-bold" style="color: rgb(175,95,135)">}</span><span class="ansi-yellow-fg">.</span><span class="ansi-yellow-fg">&#34;</span>
<span class="ansi-green-fg">    483</span>             )
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">485</span> out = <span class="ansi-yellow-bg">func</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">args</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">    486</span> <span style="color: rgb(0,135,0)">self</span>._optimizer_step_code()
<span class="ansi-green-fg">    488</span> <span style="color: rgb(95,135,135)"># call optimizer step post hooks</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">~/Desktop/package_test/.venv/lib/python3.11/site-packages/torch/optim/optimizer.py:79</span>, in <span class="ansi-cyan-fg">_use_grad_for_differentiable.&lt;locals&gt;._use_grad</span><span class="ansi-blue-fg">(self, *args, **kwargs)</span>
<span class="ansi-green-fg">     77</span>     torch.set_grad_enabled(<span style="color: rgb(0,135,0)">self</span>.defaults[<span class="ansi-yellow-fg">&#34;</span><span class="ansi-yellow-fg">differentiable</span><span class="ansi-yellow-fg">&#34;</span>])
<span class="ansi-green-fg">     78</span>     torch._dynamo.graph_break()
<span class="ansi-green-fg">---&gt; </span><span class="ansi-green-fg">79</span>     ret = <span class="ansi-yellow-bg">func</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg" style="color: rgb(0,135,0)">self</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">args</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">     80</span> <span class="ansi-bold" style="color: rgb(0,135,0)">finally</span>:
<span class="ansi-green-fg">     81</span>     torch._dynamo.graph_break()

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">~/Desktop/package_test/.venv/lib/python3.11/site-packages/torch/optim/adam.py:246</span>, in <span class="ansi-cyan-fg">Adam.step</span><span class="ansi-blue-fg">(self, closure)</span>
<span class="ansi-green-fg">    234</span>     beta1, beta2 = group[<span class="ansi-yellow-fg">&#34;</span><span class="ansi-yellow-fg">betas</span><span class="ansi-yellow-fg">&#34;</span>]
<span class="ansi-green-fg">    236</span>     has_complex = <span style="color: rgb(0,135,0)">self</span>._init_group(
<span class="ansi-green-fg">    237</span>         group,
<span class="ansi-green-fg">    238</span>         params_with_grad,
<span class="ansi-green-fg">   (...)</span><span class="ansi-green-fg">    243</span>         state_steps,
<span class="ansi-green-fg">    244</span>     )
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">246</span>     <span class="ansi-yellow-bg">adam</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg">    247</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">params_with_grad</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    248</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">grads</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    249</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">exp_avgs</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    250</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">exp_avg_sqs</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    251</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">max_exp_avg_sqs</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    252</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">state_steps</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    253</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">amsgrad</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">amsgrad</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    254</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">has_complex</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">has_complex</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    255</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">beta1</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">beta1</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    256</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">beta2</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">beta2</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    257</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">lr</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">lr</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    258</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">weight_decay</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">weight_decay</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    259</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">eps</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">eps</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    260</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">maximize</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">maximize</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    261</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">foreach</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">foreach</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    262</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">capturable</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">capturable</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    263</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">differentiable</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">differentiable</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    264</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">fused</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">fused</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    265</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">grad_scale</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg" style="color: rgb(0,135,0)">getattr</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg" style="color: rgb(0,135,0)">self</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">grad_scale</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg ansi-bold" style="color: rgb(0,135,0)">None</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    266</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">found_inf</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg" style="color: rgb(0,135,0)">getattr</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg" style="color: rgb(0,135,0)">self</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">found_inf</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg ansi-bold" style="color: rgb(0,135,0)">None</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    267</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">decoupled_weight_decay</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">group</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-fg ansi-yellow-bg">decoupled_weight_decay</span><span class="ansi-yellow-fg ansi-yellow-bg">&#34;</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    268</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">    270</span> <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> loss

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">~/Desktop/package_test/.venv/lib/python3.11/site-packages/torch/optim/optimizer.py:147</span>, in <span class="ansi-cyan-fg">_disable_dynamo_if_unsupported.&lt;locals&gt;.wrapper.&lt;locals&gt;.maybe_fallback</span><span class="ansi-blue-fg">(*args, **kwargs)</span>
<span class="ansi-green-fg">    145</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> disabled_func(*args, **kwargs)
<span class="ansi-green-fg">    146</span> <span class="ansi-bold" style="color: rgb(0,135,0)">else</span>:
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">147</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> <span class="ansi-yellow-bg">func</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">args</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">*</span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">)</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">~/Desktop/package_test/.venv/lib/python3.11/site-packages/torch/optim/adam.py:933</span>, in <span class="ansi-cyan-fg">adam</span><span class="ansi-blue-fg">(params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_steps, foreach, capturable, differentiable, fused, grad_scale, found_inf, has_complex, decoupled_weight_decay, amsgrad, beta1, beta2, lr, weight_decay, eps, maximize)</span>
<span class="ansi-green-fg">    930</span> <span class="ansi-bold" style="color: rgb(0,135,0)">else</span>:
<span class="ansi-green-fg">    931</span>     func = _single_tensor_adam
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">933</span> <span class="ansi-yellow-bg">func</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg">    934</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">params</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    935</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">grads</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    936</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">exp_avgs</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    937</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">exp_avg_sqs</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    938</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">max_exp_avg_sqs</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    939</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">state_steps</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    940</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">amsgrad</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">amsgrad</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    941</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">has_complex</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">has_complex</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    942</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">beta1</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">beta1</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    943</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">beta2</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">beta2</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    944</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">lr</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">lr</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    945</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">weight_decay</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">weight_decay</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    946</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">eps</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">eps</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    947</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">maximize</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">maximize</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    948</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">capturable</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">capturable</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    949</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">differentiable</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">differentiable</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    950</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">grad_scale</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">grad_scale</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    951</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">found_inf</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">found_inf</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    952</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">decoupled_weight_decay</span><span class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">decoupled_weight_decay</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg">    953</span> <span class="ansi-yellow-bg">)</span>

<span class="ansi-cyan-fg">File </span><span class="ansi-green-fg">~/Desktop/package_test/.venv/lib/python3.11/site-packages/torch/optim/adam.py:439</span>, in <span class="ansi-cyan-fg">_single_tensor_adam</span><span class="ansi-blue-fg">(params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_steps, grad_scale, found_inf, amsgrad, has_complex, beta1, beta2, lr, weight_decay, eps, maximize, capturable, differentiable, decoupled_weight_decay)</span>
<span class="ansi-green-fg">    436</span>     device_beta1 = beta1
<span class="ansi-green-fg">    438</span> <span style="color: rgb(95,135,135)"># Decay the first and second moment running average coefficient</span>
<span class="ansi-green-fg">--&gt; </span><span class="ansi-green-fg">439</span> <span class="ansi-yellow-bg">exp_avg</span><span class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">lerp_</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">grad</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-green-fg ansi-yellow-bg">1</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">-</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">device_beta1</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg">    441</span> <span style="color: rgb(95,135,135)"># Nested if is necessary to bypass jitscript rules</span>
<span class="ansi-green-fg">    442</span> <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> differentiable <span class="ansi-bold" style="color: rgb(175,0,255)">and</span> <span style="color: rgb(0,135,0)">isinstance</span>(beta2, Tensor):

<span class="ansi-red-fg">KeyboardInterrupt</span>:
</pre></div></div>
</div>
<p>Visualizing the aligned latent space</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_concat</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span>
    <span class="n">X</span><span class="o">=</span><span class="n">sparse</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;z_mu&#39;</span><span class="p">],</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;z_mu&#39;</span><span class="p">],</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;hamster&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;z_mu&#39;</span><span class="p">]])</span><span class="o">.</span><span class="n">toarray</span><span class="p">(),</span>
    <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type_coarse&#39;</span><span class="p">]],</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type_coarse&#39;</span><span class="p">]],</span> <span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s1">&#39;hamster&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type_coarse&#39;</span><span class="p">]]])</span>
<span class="p">)</span>

<span class="c1"># Color scheme for the liver cell dataset. Won&#39;t return nice results for other datasets.</span>
<span class="n">palette</span> <span class="o">=</span> <span class="n">return_palette</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">adata_concat</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type_coarse</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata_concat</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">)</span>
<span class="n">neighbors_workaround</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_concat</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_type_coarse&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_aligning_datasets_39_0.png" src="../_images/tutorials_aligning_datasets_39_0.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="data_preprocessing.html" class="btn btn-neutral float-left" title="Set up datasets for scSpecies" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../tips.html" class="btn btn-neutral float-right" title="Tips to fit scSpecies" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Clemens Sch√§chter.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>