

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Set up datasets for scSpecies &mdash; scspecies Documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/jupyter-sphinx.css?v=572af1d6" />
      <link rel="stylesheet" type="text/css" href="../_static/thebelab.css" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=01f34227"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/thebelab-helper.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="The scSpecies workflow" href="aligning_datasets.html" />
    <link rel="prev" title="Tutorials" href="../tutorials.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            scspecies
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Navigation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api/api.html">API Reference</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Set up datasets for scSpecies</a></li>
<li class="toctree-l2"><a class="reference internal" href="aligning_datasets.html">The scSpecies workflow</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tips.html">Tips to fit scSpecies</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scspecies</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Set up datasets for scSpecies</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/data_preprocessing.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Set-up-datasets-for-scSpecies">
<h1>Set up datasets for scSpecies<a class="headerlink" href="#Set-up-datasets-for-scSpecies" title="Link to this heading"></a></h1>
<p>This tutorial demonstrates how to setup anndata scRNA-seq datasets such that they can be integrated by scSpecies.</p>
<p>We start by downloading three liver cell datasets from mice, humans and hamsters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">download_datasets</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>

<span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\\</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;/&#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;/&#39;</span>
<span class="n">download_datasets</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
human_liver.h5ad already exists. Skipping download.
mouse_liver.h5ad already exists. Skipping download.
hamster_liver.h5ad already exists. Skipping download.
All datasets have been downloaded to the ./data directory.
</pre></div></div>
</div>
<div class="line-block">
<div class="line">As a first step we load the context and target datasets as <code class="docutils literal notranslate"><span class="pre">.h5ad</span></code> files.</div>
<div class="line">Custom datasets must have raw counts saved as <code class="docutils literal notranslate"><span class="pre">Compressed</span> <span class="pre">Sparse</span> <span class="pre">Row</span></code> sparse matrix of dtype ‘float32’.</div>
</div>
<div class="line-block">
<div class="line">As context dataset we will use the <code class="docutils literal notranslate"><span class="pre">mouse_liver.h5ad</span></code> file, which contains mouse liver cell samples.</div>
<div class="line">As target datasets we will use the <code class="docutils literal notranslate"><span class="pre">human_liver.h5ad</span></code> file, which contains human liver cell samples and the <code class="docutils literal notranslate"><span class="pre">hamster_liver.h5ad</span></code> file, which contains hamster liver cell samples.</div>
</div>
<p>NOTE: The three datasets are subsampled from the mouse, hamster and human liver cell atlas.</p>
<div class="line-block">
<div class="line">First, gene sets were reduced to 4000 highly variable genes that are expressed in more than 2.5% of cells.</div>
<div class="line">Second, cells belonging to large cell types were randomly sampled to contain 1500 samples.</div>
<div class="line">Third, unlabeled cells and cells with a labeling conflicts between fine and coarse labels were removed.</div>
<div class="line">Lastly, we only included cells obtained via CITE-seq and scRNA-seq.</div>
</div>
<p>The full datasets can be accessed via <a class="reference external" href="https://www.livercellatlas.org/">https://www.livercellatlas.org/</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scipy.sparse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sp</span>

<span class="n">adata_mouse</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s2">&quot;data/mouse_liver.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata_human</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s2">&quot;data/human_liver.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata_hamster</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">path</span><span class="o">+</span><span class="s2">&quot;data/hamster_liver.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line">Next, we specify the <code class="docutils literal notranslate"><span class="pre">.obs</span></code> keys under which the cell and batch labels for the context and target dataset are stored.</div>
<div class="line">For the target dataset, cell type annotation is not required.</div>
<div class="line">However, we will use existing annotation for visualization and to compute performance metrics. When the target cell annotation is unknown this can be indicated by <code class="docutils literal notranslate"><span class="pre">target_cell_key</span> <span class="pre">=</span> <span class="pre">None</span></code>.</div>
</div>
<p>NOTE: For precise metric calculation annotation should match across datasets.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mouse_batch_key</span> <span class="o">=</span> <span class="s1">&#39;batch&#39;</span>
<span class="n">mouse_cell_key</span> <span class="o">=</span> <span class="s1">&#39;cell_type_fine&#39;</span>

<span class="n">human_batch_key</span> <span class="o">=</span> <span class="s1">&#39;batch&#39;</span>
<span class="n">human_cell_key</span> <span class="o">=</span> <span class="s1">&#39;cell_type_fine&#39;</span>

<span class="n">hamster_batch_key</span> <span class="o">=</span> <span class="s1">&#39;batch&#39;</span>
<span class="n">hamster_cell_key</span> <span class="o">=</span> <span class="s1">&#39;cell_type_coarse&#39;</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line">Next, we specify the gene naming convention of the datasets.</div>
<div class="line">scSpecies can match homologous genes between arbitrary genomes via the NCBI Taxon ID using the package <code class="docutils literal notranslate"><span class="pre">mygene</span></code>.</div>
</div>
<p>A list of commonly used model species is provided below:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Species</p></th>
<th class="head"><p>NCBI Taxonomy ID</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>human</p></td>
<td><p>9606</p></td>
</tr>
<tr class="row-odd"><td><p>mouse</p></td>
<td><p>10090</p></td>
</tr>
<tr class="row-even"><td><p>hamster</p></td>
<td><p>10036</p></td>
</tr>
<tr class="row-odd"><td><p>rat</p></td>
<td><p>10116</p></td>
</tr>
<tr class="row-even"><td><p>zebrafish</p></td>
<td><p>7955</p></td>
</tr>
<tr class="row-odd"><td><p>fruit fly</p></td>
<td><p>7227</p></td>
</tr>
<tr class="row-even"><td><p>chicken</p></td>
<td><p>9031</p></td>
</tr>
<tr class="row-odd"><td><p>pig</p></td>
<td><p>9823</p></td>
</tr>
<tr class="row-even"><td><p>rhesus macaque</p></td>
<td><p>9544</p></td>
</tr>
<tr class="row-odd"><td><p>rabbit</p></td>
<td><p>9986</p></td>
</tr>
</tbody>
</table>
<p><em>Note: The Liver cell atlas has already mapped the gene names to mouse or human gene symbols. We will therefore set the Taxonomy ID to 10090 for the hamster dataset.</em></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mouse_NCBI_Taxon_ID</span> <span class="o">=</span> <span class="mi">10090</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">First mouse  gene names: &#39;</span><span class="p">,</span> <span class="n">adata_mouse</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">])</span>

<span class="n">human_NCBI_Taxon_ID</span> <span class="o">=</span> <span class="mi">9606</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">First human gene names: &#39;</span><span class="p">,</span> <span class="n">adata_human</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">])</span>

<span class="n">hamster_NCBI_Taxon_ID</span> <span class="o">=</span> <span class="mi">10090</span> <span class="c1">#Gene symbols follow the mouse convention in the hamster liver cell dataset.</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">First hamster gene names: &#39;</span><span class="p">,</span> <span class="n">adata_hamster</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

First mouse  gene names:  Index([&#39;Sox17&#39;, &#39;Mrpl15&#39;, &#39;Sntg1&#39;, &#39;Adhfe1&#39;, &#39;Snhg6&#39;], dtype=&#39;object&#39;)

First human gene names:  Index([&#39;RP11-54O7.17&#39;, &#39;HES4&#39;, &#39;ISG15&#39;, &#39;TNFRSF18&#39;, &#39;TNFRSF4&#39;], dtype=&#39;object&#39;)

First hamster gene names:  Index([&#39;Tcf7l2&#39;, &#39;Adra2a&#39;, &#39;Dusp5&#39;, &#39;Mxi1&#39;, &#39;Add3&#39;], dtype=&#39;object&#39;)
</pre></div></div>
</div>
<div class="line-block">
<div class="line">Next, we create a <code class="docutils literal notranslate"><span class="pre">muon.MuData</span></code> file (<a class="reference external" href="https://muon.readthedocs.io/en/latest/">https://muon.readthedocs.io/en/latest/</a>) which scSpecies uses during training.</div>
<div class="line">Each species will occupy one modality.</div>
</div>
<div class="line-block">
<div class="line">We instantiate a preprocessing class and register context and target <code class="docutils literal notranslate"><span class="pre">anndata.AnnData</span></code> datasets.</div>
<div class="line">This class translates the gene names, performs the data-level nearest neighbor search on homologous genes,</div>
<div class="line">one-hot-encodes experimental batch effects, computes the library encoder prior parameters.</div>
</div>
<div class="line-block">
<div class="line">We recommend to subset to the gene sets of interest before inputting them into the preprocessing class.</div>
<div class="line">Gene sets do not have to be of similar sizes. To demonstrate this, we randomly subset the hamster dataset to 3500 genes. The other datasets have 4000 genes.</div>
</div>
<p>NOTE: scSpecies needs integer count values without any preprocessing transformations such as log1p or normalisation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scspecies.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">create_mdata</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">UserWarning</span><span class="p">)</span>

<span class="n">preprocess</span> <span class="o">=</span> <span class="n">create_mdata</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata_mouse</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="n">mouse_batch_key</span><span class="p">,</span>
    <span class="n">cell_key</span><span class="o">=</span><span class="n">mouse_cell_key</span><span class="p">,</span>
    <span class="n">dataset_name</span><span class="o">=</span><span class="s1">&#39;mouse&#39;</span><span class="p">,</span>
    <span class="n">NCBI_Taxon_ID</span><span class="o">=</span><span class="n">mouse_NCBI_Taxon_ID</span><span class="p">,</span>
    <span class="p">)</span>

<span class="n">preprocess</span><span class="o">.</span><span class="n">setup_target_adata</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata_human</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="n">human_batch_key</span><span class="p">,</span>
    <span class="n">cell_key</span><span class="o">=</span><span class="n">human_cell_key</span><span class="p">,</span>
    <span class="n">dataset_name</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span>
    <span class="n">NCBI_Taxon_ID</span><span class="o">=</span><span class="n">human_NCBI_Taxon_ID</span><span class="p">,</span>
    <span class="p">)</span>

<span class="n">adata_hamster</span> <span class="o">=</span> <span class="n">adata_hamster</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">4000</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">3500</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)]</span>

<span class="n">preprocess</span><span class="o">.</span><span class="n">setup_target_adata</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata_hamster</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="n">hamster_batch_key</span><span class="p">,</span>
    <span class="n">cell_key</span><span class="o">=</span><span class="n">hamster_cell_key</span><span class="p">,</span>
    <span class="n">dataset_name</span><span class="o">=</span><span class="s1">&#39;hamster&#39;</span><span class="p">,</span>
    <span class="n">NCBI_Taxon_ID</span><span class="o">=</span><span class="n">mouse_NCBI_Taxon_ID</span><span class="p">,</span>
    <span class="p">)</span>

<span class="n">mdata</span> <span class="o">=</span> <span class="n">preprocess</span><span class="o">.</span><span class="n">return_mdata</span><span class="p">(</span><span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">save_name</span><span class="o">=</span><span class="s1">&#39;liver_atlas&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Registering experimental batches for the mouse dataset. Kept 34, removed 34.
Compute prior parameters for the library encoder.
Filtering cells. Kept 37717, removed 0.
Done!
------------------------------------------------------------------------------------------
Registering experimental batches for the human dataset. Kept 30, removed 30.
Compute prior parameters for the library encoder.
Filtering cells. Kept 34993, removed 0.
Translating homologous gene names between mouse context and human target dataset.
Could map <span class="ansi-yellow-fg">3535</span> of 4000 target gene symbols to context species gene symbols
Found <span class="ansi-magenta-fg">1785</span> shared homologous genes between context and target dataset
Perform the data-level nearest neigbor search on the homologous gene set.
Evaluating data level NNS and calculating cells with the highest agreement for context labels key cell_type_fine.
Done!
------------------------------------------------------------------------------------------
Registering experimental batches for the hamster dataset. Kept 2, removed 2.
Compute prior parameters for the library encoder.
Filtering cells. Kept 5953, removed 2.
Translating homologous gene names between mouse context and hamster target dataset.
Could map <span class="ansi-yellow-fg">3500</span> of 3500 target gene symbols to context species gene symbols
Found <span class="ansi-magenta-fg">1475</span> shared homologous genes between context and target dataset
Perform the data-level nearest neigbor search on the homologous gene set.
Evaluating data level NNS and calculating cells with the highest agreement for context labels key cell_type_coarse.
Done!
------------------------------------------------------------------------------------------
Saved mdata to data/liver_atlas.h5mu.
</pre></div></div>
</div>
<p>The final mdata object contains the data set names as modality keys.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">create_mdata</span></code> class has added additional keys to the <code class="docutils literal notranslate"><span class="pre">adata</span></code> input corresponding to the preprocessing steps.</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">mdata.mod[key].obs['library_log_mean']</span></code> - Library size encoder prior parameter (mu).</div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">mdata.mod[key].obs['library_log_std']</span></code> - Library size encoder prior parameter (sigma).</div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">mdata.mod[key].obs['dataset']</span></code> - Datset/species name.</div>
</div>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">mdata.mod[key].uns['metadata']</span></code> - Dataset metadata information, e.g. cell and batch annotation keys.</div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">mdata.mod[key].uns['batch_dict']</span></code>- Dictionary containing the experimental batches each cell types contains samples in.</div>
</div>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">mdata.mod[key].obsm['batch_label_enc']</span></code> - Onehot-encoded batch labels.</div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">mdata.mod[target_key].obsm['ind_nns_hom_genes']</span></code> - Indices for candidate cells of the data-level nearest neighbor search.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
MuData object with n_obs × n_vars = 78663 × 11500
  3 modalities
    mouse:      37717 x 4000
      obs:      &#39;cell_type_coarse&#39;, &#39;batch&#39;, &#39;cell_type_fine&#39;, &#39;dataset&#39;, &#39;library_log_mean&#39;, &#39;library_log_std&#39;, &#39;n_genes&#39;
      uns:      &#39;metadata&#39;, &#39;batch_dict&#39;
      obsm:     &#39;batch_label_enc&#39;
    human:      34993 x 4000
      obs:      &#39;cell_type_coarse&#39;, &#39;batch&#39;, &#39;cell_type_fine&#39;, &#39;dataset&#39;, &#39;library_log_mean&#39;, &#39;library_log_std&#39;, &#39;n_genes&#39;, &#39;top_percent_cell_type_fine&#39;, &#39;pred_nns_cell_type_fine&#39;
      var:      &#39;var_names_transl&#39;
      uns:      &#39;metadata&#39;, &#39;batch_dict&#39;
      obsm:     &#39;batch_label_enc&#39;, &#39;ind_neigh_nns&#39;
    hamster:    5953 x 3500
      obs:      &#39;cell_type_coarse&#39;, &#39;batch&#39;, &#39;dataset&#39;, &#39;library_log_mean&#39;, &#39;library_log_std&#39;, &#39;n_genes&#39;, &#39;top_percent_cell_type_coarse&#39;, &#39;pred_nns_cell_type_coarse&#39;
      var:      &#39;var_names_transl&#39;
      uns:      &#39;metadata&#39;, &#39;batch_dict&#39;
      obsm:     &#39;ind_nns_hom_genes&#39;, &#39;batch_label_enc&#39;, &#39;ind_neigh_nns&#39;
</pre></div></div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../tutorials.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="aligning_datasets.html" class="btn btn-neutral float-right" title="The scSpecies workflow" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Clemens Schächter.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>